// @carbon/ibm-products/es/components/ActionBar/ActionBarItem.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["label", "className", "renderIcon"];
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
//
// Copyright IBM Corp. 2020, 2021
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

// Import portions of React that are needed.
import React from 'react';

// Other standard imports.
import PropTypes from 'prop-types';
import cx from 'classnames';
import { pkg } from '../../settings';

// Carbon and package components we use.
import { IconButton } from '@carbon/react';

// The block part of our conventional BEM class names (blockClass__E--M).
var componentName = 'ActionBarItem';
var blockClass = "".concat(pkg.prefix, "--action-bar-item");

// NOTE: the component SCSS is not imported here: it is rolled up separately.

/**
 * The ActionBarItem is used in the page header to populate the action bar
 */
export var ActionBarItem = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
  var label = _ref.label,
    className = _ref.className,
    renderIcon = _ref.renderIcon,
    rest = _objectWithoutProperties(_ref, _excluded);
  var Icon = renderIcon;
  return /*#__PURE__*/React.createElement(IconButton, _objectSpread(_objectSpread({}, rest), {}, {
    ref: ref,
    className: cx(blockClass, className),
    kind: 'ghost',
    size: 'md',
    align: 'bottom-right',
    type: 'button',
    label: label
  }), Icon ? /*#__PURE__*/React.createElement(Icon, null) : null);
});

// Return a placeholder if not released and not enabled by feature flag
ActionBarItem = pkg.checkComponentEnabled(ActionBarItem, componentName);

// Props the user cannot change
var reservedProps = ['kind', 'size', 'align', 'type'];
// Base props on Carbon Button
var propTypes = _objectSpread({}, IconButton.propTypes);

// Remove reserved props
reservedProps.forEach(function (prop) {
  delete propTypes[prop];
});
ActionBarItem.displayName = componentName;
ActionBarItem.propTypes = _objectSpread(_objectSpread({}, propTypes), {}, {
  /* ***************************************
  /
  /  The declarations below allow storybook & DocGen to produce documentation.
  /  Some or all of them may be inherited from the underlying Carbon component.
  /
  / ****************************************/
  /**
   * Specify an optional className to be added to your Button
   *
   * (inherited from Carbon Button)
   */
  className: PropTypes.string,
  /**
   * If specifying the `renderIcon` prop, provide a description for that icon that can
   * be read by screen readers
   *
   * (inherited from Carbon Button)
   */
  label: PropTypes.string,
  /**
   * Optional click handler
   *
   * (inherited from Carbon Button)
   */
  onClick: PropTypes.func,
  /**
   * Optional prop to allow overriding the icon rendering.
   * Can be a React component class
   *
   * (inherited from Carbon Button)
   */
  renderIcon: PropTypes.oneOfType([PropTypes.func, PropTypes.object])
});
