// @carbon/ibm-products/lib/components/HTTPErrors/HTTPErrorContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HTTPErrorContent = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = require("prop-types");
var _settings = require("../../settings");
var _react2 = require("@carbon/react");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * Copyright IBM Corp. 2021, 2021
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * This source code is licensed under the Apache-2.0 license found in the
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * LICENSE file in the root directory of this source tree.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          */ // Import portions of React that are needed.
// Other standard imports.
// Carbon and package components we use.
// The block part of our conventional BEM class names (blockClass__E--M).
var blockClass = "".concat(_settings.pkg.prefix, "--http-errors");
var componentName = 'HTTPErrorContent';
var HTTPErrorContent = function HTTPErrorContent(_ref) {
  var description = _ref.description,
    errorCodeLabel = _ref.errorCodeLabel,
    title = _ref.title,
    links = _ref.links;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "__content")
  }, errorCodeLabel && /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(blockClass, "__error-code-label")
  }, errorCodeLabel), title && /*#__PURE__*/_react.default.createElement("h1", {
    className: "".concat(blockClass, "__title")
  }, title), description && /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(blockClass, "__description")
  }, description), links && links.length && links.map(function (link) {
    return /*#__PURE__*/_react.default.createElement(_react2.Link, (0, _extends2.default)({}, link, {
      role: "link",
      href: link.href,
      key: link.text,
      className: "".concat(blockClass, "__link")
    }), link.text);
  }));
};

// The display name of the component, used by React. Note that displayName
// is used in preference to relying on function.name.
exports.HTTPErrorContent = HTTPErrorContent;
HTTPErrorContent.displayName = componentName;

// The types and DocGen commentary for the component props,
// in alphabetical order (for consistency).
// See https://www.npmjs.com/package/prop-types#usage.
HTTPErrorContent.propTypes = {
  /**
   * String that will provide the description for the HTTP error code
   */
  description: _propTypes.string.isRequired,
  /**
   * String that will describe the error that occurred
   */
  errorCodeLabel: _propTypes.string.isRequired,
  /**
   * Links that will display for extra context when receiving particular errors
   */
  links: (0, _propTypes.arrayOf)((0, _propTypes.shape)(_objectSpread(_objectSpread({}, _react2.Link.propTypes), {}, {
    /**
     * The text to display for the link
     */
    text: _propTypes.string.isRequired,
    /**
     * The link's destination
     */
    href: _propTypes.string.isRequired
  }))),
  /**
   * This will be for the main title of the HTTP error component
   */
  title: _propTypes.string.isRequired
};
