// @carbon/ibm-products/lib/components/NotificationsPanel/NotificationsPanel_data.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _addonActions = require("@storybook/addon-actions");
var _uuidv = _interopRequireDefault(require("../../global/js/utils/uuidv4"));
/**
 * Copyright IBM Corp. 2020, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

var currentDate = new Date();
var yesterdayDate = new Date();
yesterdayDate.setDate(yesterdayDate.getDate() - 1);
var dayBeforeYesterday = new Date();
dayBeforeYesterday.setDate(dayBeforeYesterday.getDate() - 2);
var msInOneMinute = 60000;
var data = [{
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'LogRhythm connection failure',
  description: 'LogRhythm is failing to connect, check timeout.',
  timestamp: currentDate,
  unread: true,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'LogDNA cannot be reached.',
  description: 'Unable to communicate with LogDNA.',
  timestamp: new Date(new Date().getTime() - 30 * 1000),
  // 30 seconds ago
  unread: true,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'warning',
  title: 'System alert',
  description: 'Email classification was exported successfully.',
  timestamp: new Date(currentDate.getTime() - 11 * msInOneMinute),
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'success',
  title: 'IBM Cloud Pak for Automation Success',
  description: 'Successfully connected cartridge',
  timestamp: new Date(currentDate.getTime() - 120 * msInOneMinute),
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'success',
  title: 'Successfully connected LogDNA',
  description: 'App connection succeeded',
  timestamp: yesterdayDate,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'warning',
  title: 'Cloud Foundry app memory',
  description: 'Allocated app memory low',
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'informational',
  title: 'Logs are now being monitored',
  link: {
    text: 'View logs',
    url: 'https://www.carbondesignsystem.com'
  },
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'Cluster unreachable',
  description: 'Not able to establish connection with provided cluster. Please check your logs and memory allocation to resolve this issue further.',
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'Cluster unreachable',
  description: 'Not able to establish connection with provided cluster. Please check your logs and memory allocation to resolve this issue further.',
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'Cluster unreachable',
  description: 'Not able to establish connection with provided cluster. Please check your logs and memory allocation to resolve this issue further.',
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'Cluster unreachable',
  description: 'Not able to establish connection with provided cluster. Please check your logs and memory allocation to resolve this issue further.',
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'Cluster unreachable',
  description: 'Not able to establish connection with provided cluster. Please check your logs and memory allocation to resolve this issue further.',
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'Cluster unreachable',
  description: 'Not able to establish connection with provided cluster. Please check your logs and memory allocation to resolve this issue further.',
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'Cluster unreachable',
  description: 'Not able to establish connection with provided cluster. Please check your logs and memory allocation to resolve this issue further.',
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'Cluster unreachable',
  description: 'Not able to establish connection with provided cluster. Please check your logs and memory allocation to resolve this issue further.',
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}, {
  id: (0, _uuidv.default)(),
  type: 'error',
  title: 'Cluster unreachable',
  description: 'Not able to establish connection with provided cluster. Please check your logs and memory allocation to resolve this issue further.',
  timestamp: dayBeforeYesterday,
  onNotificationClick: (0, _addonActions.action)("Clicked on notification")
}];
var _default = data;
exports.default = _default;
