// @carbon/ibm-products/lib/components/SidePanel/motion/variants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.panelVariants = exports.overlayVariants = void 0;
var _motionConstants = require("../../../global/js/utils/motionConstants");
/**
 * Copyright IBM Corp. 2022, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

var overlayVariants = {
  visible: {
    opacity: 1,
    transition: {
      duration: _motionConstants.DURATIONS.moderate02,
      ease: _motionConstants.EASINGS.productive.standard
    }
  },
  hidden: {
    opacity: 0
  },
  exit: {
    opacity: 0
  }
};
exports.overlayVariants = overlayVariants;
var panelVariants = {
  visible: {
    x: 0,
    transition: {
      duration: _motionConstants.DURATIONS.moderate02,
      ease: _motionConstants.EASINGS.productive.standard
    }
  },
  hidden: function hidden(placement) {
    return {
      x: placement === 'right' ? '100%' : -320
    };
  },
  exit: function exit(placement) {
    return {
      x: placement === 'right' ? '100%' : -320,
      transition: {
        duration: _motionConstants.DURATIONS.moderate01,
        ease: _motionConstants.EASINGS.productive.exit
      }
    };
  }
};
exports.panelVariants = panelVariants;
