// @carbon/ibm-products/lib/components/EditTearsheet/EditTearsheet.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormNumberContext = exports.FormContext = exports.EditTearsheet = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _classnames = _interopRequireDefault(require("classnames"));
var _react2 = require("@carbon/react");
var _settings = require("../../settings");
var _devtools = require("../../global/js/utils/devtools");
var _TearsheetShell = require("../Tearsheet/TearsheetShell");
var _excluded = ["cancelButtonText", "children", "className", "description", "influencerWidth", "label", "onClose", "open", "submitButtonText", "title", "verticalPosition", "onHandleModalClick"];
/**
 * Copyright IBM Corp. 2022, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var componentName = 'EditTearsheet';
var blockClass = "".concat(_settings.pkg.prefix, "--tearsheet-edit");

// This is a general context for the forms container
// containing information about the state of the container
// and providing some callback methods for forms to use
var FormContext = /*#__PURE__*/(0, _react.createContext)(null);

// This is a context supplied separately to each form in the container
// to let it know what number it is in the sequence of forms
exports.FormContext = FormContext;
var FormNumberContext = /*#__PURE__*/(0, _react.createContext)(0);

// Default values for props
exports.FormNumberContext = FormNumberContext;
var defaults = {
  verticalPosition: 'normal',
  influencerWidth: 'narrow'
};

/**
 * Use Tearsheet with medium to complex edits. See usage guidance for further information.
 */
var EditTearsheet = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var cancelButtonText = _ref.cancelButtonText,
    children = _ref.children,
    className = _ref.className,
    description = _ref.description,
    _ref$influencerWidth = _ref.influencerWidth,
    influencerWidth = _ref$influencerWidth === void 0 ? defaults.influencerWidth : _ref$influencerWidth,
    label = _ref.label,
    onClose = _ref.onClose,
    open = _ref.open,
    submitButtonText = _ref.submitButtonText,
    title = _ref.title,
    _ref$verticalPosition = _ref.verticalPosition,
    verticalPosition = _ref$verticalPosition === void 0 ? defaults.verticalPosition : _ref$verticalPosition,
    onHandleModalClick = _ref.onHandleModalClick,
    rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var _useState = (0, _react.useState)(0),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    currentForm = _useState2[0],
    setCurrentForm = _useState2[1];
  var contentRef = (0, _react.useRef)();
  var handleCurrentForm = function handleCurrentForm(form) {
    setCurrentForm(form);
  };
  var sideNavItems = [{
    label: 'Topic Name'
  }, {
    label: 'Location'
  }, {
    label: 'Partitions'
  }, {
    label: 'Message retention'
  }];
  var influencer = /*#__PURE__*/_react.default.createElement("div", {
    className: "tearsheet-stories__dummy-influencer-block"
  }, /*#__PURE__*/_react.default.createElement(_react2.SideNav, {
    "aria-label": "Side navigation",
    className: "".concat(blockClass, "__side-nav"),
    expanded: true,
    isFixedNav: false
  }, /*#__PURE__*/_react.default.createElement(_react2.SideNavItems, null, sideNavItems.map(function (item, index) {
    return /*#__PURE__*/_react.default.createElement(_react2.SideNavMenuItem, {
      key: index,
      onClick: function onClick() {
        return handleCurrentForm(index);
      },
      isActive: currentForm === index
    }, item.label);
  }))));
  return /*#__PURE__*/_react.default.createElement(_TearsheetShell.TearsheetShell, (0, _extends2.default)({}, rest, (0, _devtools.getDevtoolsProps)(componentName), {
    actions: [{
      label: submitButtonText,
      onClick: onHandleModalClick,
      kind: 'primary'
    }, {
      label: cancelButtonText,
      onClick: onHandleModalClick,
      kind: 'secondary'
    }],
    className: (0, _classnames.default)(blockClass, className),
    description: description,
    hasCloseIcon: false,
    influencer: influencer,
    influencerPosition: "left",
    influencerWidth: influencerWidth,
    label: label,
    onClose: onClose,
    open: open,
    size: "wide",
    title: title,
    verticalPosition: verticalPosition,
    ref: ref
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "__content"),
    ref: contentRef,
    role: "main"
  }, /*#__PURE__*/_react.default.createElement(_react2.Form, null, /*#__PURE__*/_react.default.createElement(FormContext.Provider, {
    value: {
      currentForm: currentForm
    }
  }, _react.default.Children.map(children, function (child, index) {
    return /*#__PURE__*/_react.default.createElement(FormNumberContext.Provider, {
      value: index
    }, child);
  })))));
});

// Return a placeholder if not released and not enabled by feature flag
exports.EditTearsheet = EditTearsheet;
exports.EditTearsheet = EditTearsheet = _settings.pkg.checkComponentEnabled(EditTearsheet, componentName);

// The display name of the component, used by React. Note that displayName
// is used in preference to relying on function.name.
EditTearsheet.displayName = componentName;

// Note that the descriptions here should be kept in sync with those for the
// corresponding props for TearsheetNarrow and TearsheetShell components.
EditTearsheet.propTypes = {
  /**
   * The cancel button text
   */
  cancelButtonText: _propTypes.default.string,
  /**
   * The main content of the tearsheet
   */
  children: _propTypes.default.node,
  /**
   * An optional class or classes to be added to the outermost element.
   */
  className: _propTypes.default.string,
  /**
   * A description of the flow, displayed in the header area of the tearsheet.
   */
  description: _propTypes.default.node,
  /**
   * The content for the influencer section of the tearsheet, displayed
   * alongside the main content. This is typically a menu, or filter, or
   * progress indicator, or similar.
   */
  influencer: _propTypes.default.element,
  /**
   * Used to set the size of the influencer
   */
  influencerWidth: _propTypes.default.oneOf(['narrow', 'wide']),
  /**
   * A label for the tearsheet, displayed in the header area of the tearsheet
   * to maintain context for the tearsheet (e.g. as the title changes from page
   * to page of a multi-page task).
   */
  label: _propTypes.default.node,
  /**
   * An optional handler that is called when the user closes the tearsheet (by
   * clicking the close button, if enabled, or clicking outside, if enabled).
   * Returning `false` here prevents the modal from closing.
   */
  onClose: _propTypes.default.func,
  /**
   * Specifies whether the tearsheet is currently open.
   */
  onHandleModalClick: _propTypes.default.func,
  /**
   * Specifies whether the tearsheet is currently open.
   */
  open: _propTypes.default.bool,
  /**
   * The submit button text
   */
  submitButtonText: _propTypes.default.string,
  /**
   * The main title of the tearsheet, displayed in the header area.
   */
  title: _propTypes.default.node,
  /**
   * The position of the top of tearsheet in the viewport. The 'normal'
   * position (the default) is a short distance down from the top of the
   * viewport, leaving room at the top for a global header bar to show through
   * from below. The 'lower' position provides a little extra room at the top
   * to allow an action bar navigation or breadcrumbs to also show through.
   */
  verticalPosition: _propTypes.default.oneOf(['normal', 'lower'])
};
