// @carbon/ibm-products/lib/global/js/utils/motionConstants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EASINGS = exports.DURATIONS = void 0;
var _getBezierValues = require("./getBezierValues");
var _motion = require("@carbon/motion");
/**
 * Copyright IBM Corp. 2022, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

var cleanMotionValue = function cleanMotionValue(val) {
  return parseInt(val) / 1000;
};
var DURATIONS = {
  // Micro-interactions such as button and toggle
  fast01: cleanMotionValue(_motion.fast01),
  // '70ms'
  // Micro-interactions such as fade
  fast02: cleanMotionValue(_motion.fast02),
  // '110ms'
  // Micro-interactions, small expansion, short distance movements
  moderate01: cleanMotionValue(_motion.moderate01),
  // '150ms'
  // Expansion, system communication, toast
  moderate02: cleanMotionValue(_motion.moderate02),
  // '240ms'
  // Large expansion, important system notifications
  slow01: cleanMotionValue(_motion.slow01),
  // '400ms'
  //  Background dimming
  slow02: cleanMotionValue(_motion.slow02) // '700ms'
};
exports.DURATIONS = DURATIONS;
var EASINGS = {
  productive: {
    entrance: (0, _getBezierValues.getBezierValues)('entrance', 'productive'),
    standard: (0, _getBezierValues.getBezierValues)('standard', 'productive'),
    exit: (0, _getBezierValues.getBezierValues)('exit', 'productive')
  },
  expressive: {
    entrance: (0, _getBezierValues.getBezierValues)('entrance', 'expressive'),
    standard: (0, _getBezierValues.getBezierValues)('standard', 'expressive'),
    exit: (0, _getBezierValues.getBezierValues)('exit', 'expressive')
  }
};
exports.EASINGS = EASINGS;
