// @carbon/ibm-products/lib/components/Datagrid/Datagrid/DatagridToolbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _icons = require("@carbon/react/icons");
var _react2 = require("@carbon/react");
var _useResizeObserver2 = require("../../../global/js/hooks/useResizeObserver");
var _ButtonMenu = require("../../ButtonMenu");
var _settings = require("../../../settings");
var _classnames = _interopRequireDefault(require("classnames"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * Copyright IBM Corp. 2022, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

var blockClass = "".concat(_settings.pkg.prefix, "--datagrid");
var DatagridBatchActionsToolbar = function DatagridBatchActionsToolbar(datagridState, width, ref) {
  var _Object$keys;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    displayAllInMenu = _useState2[0],
    setDisplayAllInMenu = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    initialListWidth = _useState4[0],
    setInitialListWidth = _useState4[1];
  var _useState5 = (0, _react.useState)(false),
    _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
    receivedInitialWidth = _useState6[0],
    setReceivedInitialWidth = _useState6[1];
  var selectedRowIds = datagridState.state.selectedRowIds,
    toggleAllRowsSelected = datagridState.toggleAllRowsSelected,
    toolbarBatchActions = datagridState.toolbarBatchActions,
    setGlobalFilter = datagridState.setGlobalFilter,
    rows = datagridState.rows;
  var totalSelected = (_Object$keys = Object.keys(selectedRowIds || {})) === null || _Object$keys === void 0 ? void 0 : _Object$keys.length;

  // Get initial width of batch actions container,
  // used to measure when all items are put inside
  // the ButtonMenu
  (0, _react.useEffect)(function () {
    if (totalSelected === 1 && !receivedInitialWidth) {
      var batchActionListWidth = ref.current.querySelector(".".concat(_settings.carbon.prefix, "--action-list")).offsetWidth;
      setInitialListWidth(batchActionListWidth);
      setReceivedInitialWidth(true);
    }
  }, [totalSelected, receivedInitialWidth, ref]);
  (0, _react.useEffect)(function () {
    var summaryWidth = ref.current.querySelector(".".concat(_settings.carbon.prefix, "--batch-summary")).offsetWidth;
    if (width < summaryWidth + initialListWidth + 32) {
      setDisplayAllInMenu(true);
    } else {
      setDisplayAllInMenu(false);
    }
  }, [width, ref, initialListWidth]);
  var getSelectedRowData = function getSelectedRowData() {
    var selectedRowIndexes = Object.keys(selectedRowIds);
    var selectedRowData = selectedRowIndexes && selectedRowIndexes.length ? selectedRowIndexes.map(function (rowIndex) {
      var filteredRow = rows.filter(function (row) {
        return row.index === parseInt(rowIndex);
      });
      if (filteredRow.length) {
        return filteredRow[0];
      }
      return [];
    }) : [];
    return selectedRowData;
  };

  // Render batch actions in ButtonMenu
  var renderBatchActionOverflow = function renderBatchActionOverflow() {
    var minWidthBeforeOverflowIcon = 380;
    // Do not render ButtonMenu when there are 3 or less items
    // and if there is enough available space to render all the items
    if ((toolbarBatchActions === null || toolbarBatchActions === void 0 ? void 0 : toolbarBatchActions.length) <= 3 && !displayAllInMenu) {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement(_ButtonMenu.ButtonMenu, {
      label: width > minWidthBeforeOverflowIcon ? 'More' : null,
      renderIcon: width > minWidthBeforeOverflowIcon ? _icons.Add : _icons.OverflowMenuVertical,
      className: (0, _classnames.default)("".concat(blockClass, "__button-menu"), (0, _defineProperty2.default)({}, "".concat(blockClass, "__button-menu--icon-only"), width <= minWidthBeforeOverflowIcon)),
      menuOptionsClass: "".concat(blockClass, "__button-menu-options"),
      flipped: true,
      menuAriaLabel: 'Batch actions'
    }, toolbarBatchActions && toolbarBatchActions.map(function (batchAction, index) {
      if (index < 2) {
        if (displayAllInMenu) {
          return /*#__PURE__*/_react.default.createElement(_ButtonMenu.ButtonMenuItem, {
            key: "".concat(batchAction.label, "-").concat(index),
            itemText: batchAction.label,
            onClick: function onClick(event) {
              batchAction.onClick(getSelectedRowData(), event);
              if (batchAction.type === 'select_all') {
                toggleAllRowsSelected(true);
              }
            }
          });
        }
        return null;
      }
      return /*#__PURE__*/_react.default.createElement(_ButtonMenu.ButtonMenuItem, {
        key: "".concat(batchAction.label, "-").concat(index),
        itemText: batchAction.label,
        onClick: function onClick(event) {
          batchAction.onClick(getSelectedRowData(), event);
          if (batchAction.type === 'select_all') {
            toggleAllRowsSelected(true);
          }
        }
      });
    }));
  };

  // Only display the first two batch actions, the rest are
  // displayed inside of the ButtonMenu if there are more than
  // 3 batch actions
  return /*#__PURE__*/_react.default.createElement(_react2.TableBatchActions, {
    shouldShowBatchActions: totalSelected > 0,
    totalSelected: totalSelected,
    onCancel: function onCancel() {
      toggleAllRowsSelected(false);
      setGlobalFilter(null);
    }
  }, !displayAllInMenu && toolbarBatchActions && (toolbarBatchActions === null || toolbarBatchActions === void 0 ? void 0 : toolbarBatchActions.map(function (batchAction, index) {
    if (index < 2 && toolbarBatchActions.length > 3 || index < 3 && toolbarBatchActions.length <= 3) {
      return /*#__PURE__*/_react.default.createElement(_react2.TableBatchAction, {
        key: "".concat(batchAction.label, "-").concat(index),
        renderIcon: batchAction.renderIcon,
        onClick: function onClick(event) {
          batchAction.onClick(getSelectedRowData(), event);
          if (batchAction.type === 'select_all') {
            toggleAllRowsSelected(true);
          }
        },
        iconDescription: batchAction.label
      }, batchAction.label);
    }
  })), renderBatchActionOverflow());
};
var DatagridToolbar = function DatagridToolbar(datagridState) {
  var ref = (0, _react.useRef)(null);
  var _useResizeObserver = (0, _useResizeObserver2.useResizeObserver)(ref),
    width = _useResizeObserver.width;
  var DatagridActions = datagridState.DatagridActions,
    DatagridBatchActions = datagridState.DatagridBatchActions,
    batchActions = datagridState.batchActions,
    rowSize = datagridState.rowSize;
  var getRowHeight = rowSize ? rowSize : 'lg';
  return batchActions && DatagridActions ? /*#__PURE__*/_react.default.createElement("div", {
    ref: ref,
    className: (0, _classnames.default)("".concat(blockClass, "__table-toolbar"), "".concat(blockClass, "__table-toolbar--").concat(getRowHeight))
  }, /*#__PURE__*/_react.default.createElement(_react2.TableToolbar, null, DatagridActions && DatagridActions(datagridState), DatagridBatchActionsToolbar && DatagridBatchActionsToolbar(datagridState, width, ref))) : DatagridActions ? /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "__table-toolbar")
  }, /*#__PURE__*/_react.default.createElement(_react2.TableToolbar, null, DatagridActions && DatagridActions(datagridState), DatagridBatchActions && DatagridBatchActions(datagridState))) : null;
};
var _default = DatagridToolbar;
exports.default = _default;
