// @carbon/ibm-products/lib/components/Datagrid/Datagrid/DatagridEmptyBody.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _settings = require("../../../settings");
var _react2 = require("@carbon/react");
var _EmptyStates = require("../../EmptyStates");
/**
 * Copyright IBM Corp. 2020, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

var blockClass = "".concat(_settings.pkg.prefix, "--datagrid");
var DatagridEmptyBody = function DatagridEmptyBody(datagridState) {
  var getTableBodyProps = datagridState.getTableBodyProps,
    headers = datagridState.headers,
    emptyStateTitle = datagridState.emptyStateTitle,
    emptyStateDescription = datagridState.emptyStateDescription,
    emptyStateSize = datagridState.emptyStateSize,
    _datagridState$emptyS = datagridState.emptyStateType,
    emptyStateType = _datagridState$emptyS === void 0 ? 'noData' : _datagridState$emptyS,
    illustrationTheme = datagridState.illustrationTheme,
    emptyStateAction = datagridState.emptyStateAction,
    emptyStateLink = datagridState.emptyStateLink;
  var emptyStateProps = {
    illustrationTheme: illustrationTheme,
    size: emptyStateSize,
    title: emptyStateTitle,
    subtitle: emptyStateDescription,
    action: emptyStateAction,
    link: emptyStateLink
  };
  return /*#__PURE__*/_react.default.createElement(_react2.TableBody, (0, _extends2.default)({}, getTableBodyProps({
    role: false
  }), {
    className: "".concat(blockClass, "__empty-state-body")
  }), /*#__PURE__*/_react.default.createElement(_react2.TableRow, null, /*#__PURE__*/_react.default.createElement(_react2.TableCell, {
    colSpan: headers.length
  }, emptyStateType === 'error' && /*#__PURE__*/_react.default.createElement(_EmptyStates.ErrorEmptyState, emptyStateProps), emptyStateType === 'noData' && /*#__PURE__*/_react.default.createElement(_EmptyStates.NoDataEmptyState, emptyStateProps), emptyStateType === 'notFound' && /*#__PURE__*/_react.default.createElement(_EmptyStates.NotFoundEmptyState, emptyStateProps))));
};
var _default = DatagridEmptyBody;
exports.default = _default;
