// @carbon/ibm-products/lib/components/Datagrid/Datagrid/DatagridVirtualBody.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _reactWindow = require("react-window");
var _react2 = require("@carbon/react");
var _settings = require("../../../settings");
var _DatagridHead = _interopRequireDefault(require("./DatagridHead"));
var _layout = require("@carbon/layout");
var _useResizeObserver = require("../../../global/js/hooks/useResizeObserver");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; } /*
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * Licensed Materials - Property of IBM
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * 5724-Q36
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * (c) Copyright IBM Corp. 2020 - 2023
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * US Government Users Restricted Rights - Use, duplication or disclosure
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * restricted by GSA ADP Schedule Contract with IBM Corp.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          */
var blockClass = "".concat(_settings.pkg.prefix, "--datagrid");
var rowSizeMap = {
  xs: 24,
  sm: 32,
  md: 40,
  lg: 48,
  xl: 64,
  // TODO: deprecate the below values in next major release (v8) on carbon-components-react
  short: 32,
  compact: 24,
  normal: 48,
  tall: 64
};
var defaultRowHeight = rowSizeMap.lg;
var DatagridVirtualBody = function DatagridVirtualBody(datagridState) {
  var _gridRef$current, _gridRef$current2;
  var getTableBodyProps = datagridState.getTableBodyProps,
    rows = datagridState.rows,
    prepareRow = datagridState.prepareRow,
    onScroll = datagridState.onScroll,
    innerListRef = datagridState.innerListRef,
    _datagridState$tableH = datagridState.tableHeight,
    tableHeight = _datagridState$tableH === void 0 ? 400 : _datagridState$tableH,
    virtualHeight = datagridState.virtualHeight,
    listRef = datagridState.listRef,
    rowSize = datagridState.rowSize,
    DatagridPagination = datagridState.DatagridPagination,
    page = datagridState.page,
    handleResize = datagridState.handleResize,
    gridRef = datagridState.gridRef;
  var handleVirtualGridResize = function handleVirtualGridResize() {
    var gridRefElement = gridRef === null || gridRef === void 0 ? void 0 : gridRef.current;
    gridRefElement.style.width = gridRefElement === null || gridRefElement === void 0 ? void 0 : gridRefElement.clientWidth;
  };
  (0, _useResizeObserver.useResizeObserver)(gridRef, handleVirtualGridResize);
  var syncScroll = function syncScroll(e) {
    var virtualBody = e.target;
    document.querySelector(".".concat(blockClass, "__head-wrap")).scrollLeft = virtualBody.scrollLeft;
    var spacerColumn = document.querySelector(".".concat(blockClass, "__head-wrap thead th:last-child"));
    spacerColumn.style.width = (0, _layout.px)(32 + (virtualBody.offsetWidth - virtualBody.clientWidth)); // scrollbar width to header column to fix header alignment
  };

  (0, _react.useEffect)(function () {
    handleResize();
  }, [handleResize]);
  var rowHeight = rowSize && rowSizeMap[rowSize] || defaultRowHeight;
  if (listRef && listRef.current) {
    listRef.current.resetAfterIndex(0);
  }
  var visibleRows = DatagridPagination && page || rows;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "__head-wrap"),
    style: {
      width: (_gridRef$current = gridRef.current) === null || _gridRef$current === void 0 ? void 0 : _gridRef$current.clientWidth,
      overflow: 'hidden'
    }
  }, /*#__PURE__*/_react.default.createElement(_DatagridHead.default, datagridState)), /*#__PURE__*/_react.default.createElement(_react2.TableBody, (0, _extends2.default)({}, getTableBodyProps({
    role: false
  }), {
    onScroll: function onScroll(e) {
      return syncScroll(e);
    }
  }), /*#__PURE__*/_react.default.createElement(_reactWindow.VariableSizeList, {
    height: virtualHeight || tableHeight,
    itemCount: visibleRows.length,
    itemSize: function itemSize(index) {
      return visibleRows[index].isExpanded ? (visibleRows[index].expandedContentHeight || 0) + rowHeight : rowHeight;
    },
    estimatedItemSize: rowHeight,
    onScroll: onScroll,
    innerRef: innerListRef,
    ref: listRef,
    className: "".concat(blockClass, "__virtual-scrollbar"),
    style: {
      width: (_gridRef$current2 = gridRef.current) === null || _gridRef$current2 === void 0 ? void 0 : _gridRef$current2.clientWidth
    }
  }, function (_ref) {
    var index = _ref.index,
      style = _ref.style;
    var row = visibleRows[index];
    prepareRow(row);
    return /*#__PURE__*/_react.default.createElement("div", {
      style: _objectSpread({}, style)
    }, row.RowRenderer(_objectSpread(_objectSpread({}, datagridState), {}, {
      row: row
    })));
  })));
};
var _default = DatagridVirtualBody;
exports.default = _default;
