// @carbon/ibm-products/lib/components/Datagrid/Datagrid/addons/Filtering/motion/variants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.panelVariants = exports.innerContainerVariants = exports.actionSetVariants = void 0;
var _motionConstants = require("../../../../../../global/js/utils/motionConstants");
var _constants = require("../constants");
var panelVariants = {
  hidden: {
    width: 0,
    overflow: 'hidden',
    transition: {
      duration: _motionConstants.DURATIONS.fast02,
      ease: _motionConstants.EASINGS.productive.exit,
      when: 'afterChildren'
    }
  },
  visible: {
    width: _constants.PANEL_WIDTH,
    overflow: 'visible',
    transition: {
      duration: _motionConstants.DURATIONS.moderate02,
      ease: _motionConstants.EASINGS.productive.entrance,
      when: 'beforeChildren'
    }
  }
};
exports.panelVariants = panelVariants;
var innerContainerVariants = {
  hidden: {
    opacity: 0,
    transition: {
      duration: _motionConstants.DURATIONS.fast01,
      ease: _motionConstants.EASINGS.productive.exit
    }
  },
  visible: {
    opacity: 1,
    transition: {
      duration: _motionConstants.DURATIONS.fast02,
      ease: _motionConstants.EASINGS.productive.entrance,
      when: 'beforeChildren'
    }
  }
};
exports.innerContainerVariants = innerContainerVariants;
var actionSetVariants = {
  hidden: {
    y: _constants.ACTION_SET_HEIGHT,
    transition: {
      duration: _motionConstants.DURATIONS.fast01,
      ease: _motionConstants.EASINGS.productive.exit
    }
  },
  visible: {
    y: 0,
    transition: {
      duration: _motionConstants.DURATIONS.fast02,
      ease: _motionConstants.EASINGS.productive.entrance
    }
  }
};
exports.actionSetVariants = actionSetVariants;
