// @carbon/ibm-products/es/components/Datagrid/Datagrid/addons/CustomizeColumns/Actions.js
/**
 * Copyright IBM Corp. 2022, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import PropTypes from 'prop-types';
import { Search } from '@carbon/react';
import * as React from 'react';
import { pkg } from '../../../../../settings';
var blockClass = "".concat(pkg.prefix, "--datagrid");
var Actions = function Actions(_ref) {
  var searchText = _ref.searchText,
    setSearchText = _ref.setSearchText,
    findColumnPlaceholderLabel = _ref.findColumnPlaceholderLabel;
  return /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "__customize-columns-tearsheet--actions")
  }, /*#__PURE__*/React.createElement(Search, {
    placeholder: findColumnPlaceholderLabel,
    value: searchText,
    size: "lg",
    labelText: findColumnPlaceholderLabel,
    onChange: function onChange(e) {
      // TODO: is it performant?
      setSearchText(e.target.value);
    }
  }));
};
Actions.propTypes = {
  columns: PropTypes.array.isRequired,
  findColumnPlaceholderLabel: PropTypes.string,
  originalColumnDefinitions: PropTypes.array.isRequired,
  resetToDefaultLabel: PropTypes.string,
  searchText: PropTypes.string.isRequired,
  setColumnsObject: PropTypes.func.isRequired,
  setSearchText: PropTypes.func.isRequired
};
export default Actions;
