// @carbon/ibm-products/lib/components/ComboButton/ComboButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComboButton = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _toArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toArray"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _icons = require("@carbon/react/icons");
var _react = require("@carbon/react");
var _classnames = _interopRequireDefault(require("classnames"));
var _propTypes = require("prop-types");
var _uuidv = _interopRequireDefault(require("../../global/js/utils/uuidv4"));
var _react2 = _interopRequireWildcard(require("react"));
var _excluded = ["children", "className", "overflowMenu"],
  _excluded2 = ["children"],
  _excluded3 = ["children", "renderIcon"];
/**
 * Copyright IBM Corp. 2020, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var blockClass = 'security--combo-button';

/**
 * The combo button consolidates similar actions, while exposing the most commonly used one.
 */
var ComboButton = function ComboButton(_ref) {
  var children = _ref.children,
    className = _ref.className,
    overflowMenu = _ref.overflowMenu,
    rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var _useRef = (0, _react2.useRef)((0, _uuidv.default)()),
    instanceId = _useRef.current;
  var _useState = (0, _react2.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    isOpen = _useState2[0],
    setIsOpen = _useState2[1];
  var _Children$toArray$fil = _react2.Children.toArray(children).filter(Boolean).map(function (_ref2) {
      var _ref2$props = _ref2.props,
        children = _ref2$props.children,
        props = (0, _objectWithoutProperties2.default)(_ref2$props, _excluded2);
      return _objectSpread(_objectSpread({}, props), {}, {
        children: /*#__PURE__*/_react2.default.createElement("span", {
          className: "".concat(blockClass, "__action"),
          title: children
        }, children)
      });
    }),
    _Children$toArray$fil2 = (0, _toArray2.default)(_Children$toArray$fil),
    primaryAction = _Children$toArray$fil2[0],
    restActions = _Children$toArray$fil2.slice(1);
  return /*#__PURE__*/_react2.default.createElement("div", (0, _extends2.default)({}, rest, {
    className: (0, _classnames.default)(blockClass, className),
    "data-floating-menu-container": true
  }), /*#__PURE__*/_react2.default.createElement(_react.Button, primaryAction), restActions.length > 0 && /*#__PURE__*/_react2.default.createElement(_react.OverflowMenu, (0, _extends2.default)({}, overflowMenu, {
    className: "".concat(blockClass, "__overflow-menu"),
    menuOptionsClass: "".concat(blockClass, "__overflow-menu__list"),
    onClick: function onClick() {
      return !isOpen && setIsOpen(true);
    },
    onClose: function onClose() {
      return setIsOpen(false);
    },
    renderIcon: function renderIcon() {
      return /*#__PURE__*/(0, _react2.createElement)(isOpen ? function (props) {
        return /*#__PURE__*/_react2.default.createElement(_icons.ChevronUp, (0, _extends2.default)({
          size: 16
        }, props));
      } : function (props) {
        return /*#__PURE__*/_react2.default.createElement(_icons.ChevronDown, (0, _extends2.default)({
          size: 16
        }, props));
      }, {
        className: "".concat(blockClass, "__overflow-menu__icon")
      });
    },
    flipped: true
  }), restActions.map(function (_ref3, index) {
    var children = _ref3.children,
      Icon = _ref3.renderIcon,
      action = (0, _objectWithoutProperties2.default)(_ref3, _excluded3);
    return /*#__PURE__*/_react2.default.createElement(_react.OverflowMenuItem, (0, _extends2.default)({}, action, {
      key: "".concat(blockClass, "--").concat(instanceId, "__overflow-menu__item__").concat(index),
      className: "".concat(blockClass, "__overflow-menu__item"),
      itemText: /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, children, Icon && /*#__PURE__*/_react2.default.createElement("span", {
        className: "".concat(blockClass, "__overflow-menu__item__icon")
      }, /*#__PURE__*/_react2.default.createElement(Icon, null)))
    }));
  })));
};
exports.ComboButton = ComboButton;
ComboButton.propTypes = {
  /** Provide the contents of the `ComboButton` */
  children: _propTypes.node.isRequired,
  /** Provide an optional class to be applied to the containing node */
  className: _propTypes.string,
  /** Provide the [props of the `OverflowMenu`](https://react.carbondesignsystem.com/?path=/docs/overflowmenu) */
  overflowMenu: (0, _propTypes.shape)(_react.OverflowMenu.propTypes)
};
