// @carbon/ibm-products/lib/components/AddSelect/hooks/usePath.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = require("react");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; } //
// Copyright IBM Corp. 2022, 2022
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//
var usePath = function usePath() {
  var itemsLabel = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  var _useState = (0, _react.useState)([{
      id: 'base_of_path',
      title: itemsLabel
    }]),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    path = _useState2[0],
    setPath = _useState2[1];
  (0, _react.useEffect)(function () {
    setPath([{
      id: 'base_of_path',
      title: itemsLabel
    }]);
  }, [itemsLabel]);
  var handler = function handler(id, title, parentId) {
    if (path.find(function (entry) {
      return entry.id === id;
    })) {
      return;
    }
    var pathEntry = _objectSpread({
      id: id,
      title: title
    }, parentId && {
      parentId: parentId
    });
    if (path.find(function (entry) {
      return entry.parentId === parentId;
    })) {
      var parentIdx = path.findIndex(function (entry) {
        return entry.id === parentId;
      });
      var pathCopy = (0, _toConsumableArray2.default)(path);
      pathCopy.length = parentIdx + 1;
      setPath([].concat((0, _toConsumableArray2.default)(pathCopy), [pathEntry]));
    } else {
      setPath([].concat((0, _toConsumableArray2.default)(path), [pathEntry]));
    }
  };
  var pathClickHandler = function pathClickHandler(idx) {
    var pathCopy = (0, _toConsumableArray2.default)(path);
    pathCopy.length = idx + 1;
    setPath((0, _toConsumableArray2.default)(pathCopy));
  };
  var resetPath = function resetPath() {
    setPath([{
      id: 'base_of_path',
      title: itemsLabel
    }]);
  };
  return {
    path: path,
    setPath: handler,
    pathOnclick: pathClickHandler,
    resetPath: resetPath
  };
};
var _default = usePath;
exports.default = _default;
