// @carbon/ibm-products/es/components/Datagrid/Datagrid/DatagridEmptyBody.js
import _extends from "@babel/runtime/helpers/extends";
/**
 * Copyright IBM Corp. 2020, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { pkg } from '../../../settings';
import { TableBody, TableRow, TableCell } from '@carbon/react';
import { NoDataEmptyState, ErrorEmptyState, NotFoundEmptyState } from '../../EmptyStates';
var blockClass = "".concat(pkg.prefix, "--datagrid");
var DatagridEmptyBody = function DatagridEmptyBody(datagridState) {
  var getTableBodyProps = datagridState.getTableBodyProps,
    headers = datagridState.headers,
    emptyStateTitle = datagridState.emptyStateTitle,
    emptyStateDescription = datagridState.emptyStateDescription,
    emptyStateSize = datagridState.emptyStateSize,
    _datagridState$emptyS = datagridState.emptyStateType,
    emptyStateType = _datagridState$emptyS === void 0 ? 'noData' : _datagridState$emptyS,
    illustrationTheme = datagridState.illustrationTheme,
    emptyStateAction = datagridState.emptyStateAction,
    emptyStateLink = datagridState.emptyStateLink;
  var emptyStateProps = {
    illustrationTheme: illustrationTheme,
    size: emptyStateSize,
    title: emptyStateTitle,
    subtitle: emptyStateDescription,
    action: emptyStateAction,
    link: emptyStateLink
  };
  return /*#__PURE__*/React.createElement(TableBody, _extends({}, getTableBodyProps({
    role: false
  }), {
    className: "".concat(blockClass, "__empty-state-body")
  }), /*#__PURE__*/React.createElement(TableRow, null, /*#__PURE__*/React.createElement(TableCell, {
    colSpan: headers.length
  }, emptyStateType === 'error' && /*#__PURE__*/React.createElement(ErrorEmptyState, emptyStateProps), emptyStateType === 'noData' && /*#__PURE__*/React.createElement(NoDataEmptyState, emptyStateProps), emptyStateType === 'notFound' && /*#__PURE__*/React.createElement(NotFoundEmptyState, emptyStateProps))));
};
export default DatagridEmptyBody;
