// @carbon/ibm-products/es/components/Datagrid/Datagrid/DatagridRow.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
var _excluded = ["children"];
/*
 * Licensed Materials - Property of IBM
 * 5724-Q36
 * (c) Copyright IBM Corp. 2020
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 */
// @flow
import React from 'react';
import { TableRow, TableCell, SkeletonText } from '@carbon/react';
import { px } from '@carbon/layout';
import { selectionColumnId } from '../common-column-ids';
import cx from 'classnames';
import { pkg, carbon } from '../../../settings';
var blockClass = "".concat(pkg.prefix, "--datagrid");
var rowHeights = {
  xs: 24,
  sm: 32,
  md: 40,
  lg: 48,
  xl: 64
};

// eslint-disable-next-line react/prop-types
var DatagridRow = function DatagridRow(datagridState) {
  var _cx;
  var row = datagridState.row,
    rowSize = datagridState.rowSize,
    withNestedRows = datagridState.withNestedRows;
  var getVisibleNestedRowCount = function getVisibleNestedRowCount(_ref) {
    var isExpanded = _ref.isExpanded,
      subRows = _ref.subRows;
    var size = 0;
    if (isExpanded && subRows) {
      size += subRows.length;
      subRows.forEach(function (child) {
        size += getVisibleNestedRowCount(child);
      });
    }
    return size;
  };
  var hoverHandler = function hoverHandler(event) {
    var _hoverRow$style, _hoverRow$style2, _hoverRow$style3;
    var subRowCount = getVisibleNestedRowCount(row);
    var totalNestedRowIndicatorHeight = px(subRowCount * rowHeights[rowSize]);
    var hoverRow = event.target.closest(".".concat(blockClass, "__carbon-row-expanded"));
    hoverRow === null || hoverRow === void 0 ? void 0 : hoverRow.classList.add("".concat(blockClass, "__carbon-row-expanded-hover-active"));
    var rowExpanderButton = hoverRow === null || hoverRow === void 0 ? void 0 : hoverRow.querySelector(".".concat(blockClass, "__row-expander"));
    var rowSizeValue = rowSize || 'lg';
    hoverRow === null || hoverRow === void 0 ? void 0 : (_hoverRow$style = hoverRow.style) === null || _hoverRow$style === void 0 ? void 0 : _hoverRow$style.setProperty("--".concat(blockClass, "--indicator-height"), totalNestedRowIndicatorHeight);
    hoverRow === null || hoverRow === void 0 ? void 0 : (_hoverRow$style2 = hoverRow.style) === null || _hoverRow$style2 === void 0 ? void 0 : _hoverRow$style2.setProperty("--".concat(blockClass, "--row-height"), px(rowHeights[rowSizeValue]));
    hoverRow === null || hoverRow === void 0 ? void 0 : (_hoverRow$style3 = hoverRow.style) === null || _hoverRow$style3 === void 0 ? void 0 : _hoverRow$style3.setProperty("--".concat(blockClass, "--indicator-offset-amount"), px((rowExpanderButton === null || rowExpanderButton === void 0 ? void 0 : rowExpanderButton.offsetLeft) || 0));
  };
  var focusRemover = function focusRemover() {
    var elements = document.querySelectorAll(".".concat(blockClass, "__carbon-row-expanded"));
    elements.forEach(function (el) {
      el.classList.remove("".concat(blockClass, "__carbon-row-expanded-hover-active"));
    });
  };
  return /*#__PURE__*/React.createElement(TableRow, _extends({
    className: cx("".concat(blockClass, "__carbon-row"), (_cx = {}, _defineProperty(_cx, "".concat(blockClass, "__carbon-row-expanded"), row.isExpanded), _defineProperty(_cx, "".concat(blockClass, "__carbon-row-expandable"), row.canExpand), _defineProperty(_cx, "".concat(carbon.prefix, "--data-table--selected"), row.isSelected), _cx))
  }, row.getRowProps({
    role: false
  }), {
    key: row.id,
    onMouseEnter: function onMouseEnter(event) {
      if (!withNestedRows) {
        return;
      }
      hoverHandler(event);
    },
    onMouseLeave: function onMouseLeave(event) {
      var hoverRow = event.target.closest(".".concat(blockClass, "__carbon-row-expanded"));
      hoverRow === null || hoverRow === void 0 ? void 0 : hoverRow.classList.remove("".concat(blockClass, "__carbon-row-expanded-hover-active"));
    },
    onFocus: function onFocus(event) {
      if (!withNestedRows) {
        return;
      }
      hoverHandler(event);
    },
    onBlur: function onBlur() {
      focusRemover();
    },
    onKeyUp: function onKeyUp(event) {
      if (!withNestedRows) {
        return;
      }
      if (event.key === 'Enter' || event.key === 'Space') {
        focusRemover();
        hoverHandler(event);
      }
    }
  }), row.cells.map(function (cell, index) {
    var cellProps = cell.getCellProps({
      role: false
    });
    var children = cellProps.children,
      restProps = _objectWithoutProperties(cellProps, _excluded);
    var content = children || /*#__PURE__*/React.createElement(React.Fragment, null, !row.isSkeleton && cell.render('Cell'), row.isSkeleton && /*#__PURE__*/React.createElement(SkeletonText, null));
    if (cell && cell.column && cell.column.id === selectionColumnId) {
      // directly render component without the wrapping TableCell
      return cell.render('Cell', {
        key: cell.column.id
      });
    }
    return /*#__PURE__*/React.createElement(TableCell, _extends({
      className: cx("".concat(blockClass, "__cell"), _defineProperty({}, "".concat(blockClass, "__expandable-row-cell"), row.canExpand && index === 0))
    }, restProps, {
      key: cell.column.id
    }), content);
  }));
};
export default DatagridRow;
