// @carbon/ibm-products/lib/components/Datagrid/Datagrid/addons/Filtering/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RADIO = exports.PANEL_WIDTH = exports.PANEL = exports.NUMBER = exports.INSTANT = exports.FLYOUT = exports.DROPDOWN = exports.DATE = exports.CLEAR_FILTERS = exports.CHECKBOX = exports.BATCH = exports.ACTION_SET_HEIGHT = void 0;
/**
 * Copyright IBM Corp. 2022, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
/** Constants for update methods can either be batch or instant */
var BATCH = 'batch';
exports.BATCH = BATCH;
var INSTANT = 'instant';

/** Constants for filter variation */
exports.INSTANT = INSTANT;
var FLYOUT = 'flyout';
exports.FLYOUT = FLYOUT;
var PANEL = 'panel';

/** Constants for filter type */
exports.PANEL = PANEL;
var DATE = 'date';
exports.DATE = DATE;
var NUMBER = 'number';
exports.NUMBER = NUMBER;
var CHECKBOX = 'checkbox';
exports.CHECKBOX = CHECKBOX;
var RADIO = 'radio';
exports.RADIO = RADIO;
var DROPDOWN = 'dropdown';

/** Constants for event emitters */
exports.DROPDOWN = DROPDOWN;
var CLEAR_FILTERS = 'clearFilters';

/** Constants for panel dimensions */
exports.CLEAR_FILTERS = CLEAR_FILTERS;
var PANEL_WIDTH = 320;
exports.PANEL_WIDTH = PANEL_WIDTH;
var ACTION_SET_HEIGHT = 64;
exports.ACTION_SET_HEIGHT = ACTION_SET_HEIGHT;
