// @carbon/ibm-products/lib/components/AddSelect/AddSelectSort.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AddSelectSort = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react2 = require("@carbon/react");
var _icons = require("@carbon/react/icons");
var _settings = require("../../settings");
//
// Copyright IBM Corp. 2022, 2023
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

var blockClass = "".concat(_settings.pkg.prefix, "--add-select-sort");
var componentName = 'AddSelectSort';
var AddSelectSort = function AddSelectSort(_ref) {
  var setSortAttribute = _ref.setSortAttribute,
    setSortDirection = _ref.setSortDirection,
    sortBy = _ref.sortBy,
    sortByLabel = _ref.sortByLabel;
  var sortByOpts = sortBy ? sortBy.reduce(function (acc, cur) {
    var opts = [{
      id: "".concat(cur, "-asc"),
      itemText: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_icons.ArrowUp, {
        size: 16
      }), cur),
      direction: 'asc',
      attribute: cur
    }, {
      id: "".concat(cur, "-desc"),
      itemText: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_icons.ArrowDown, {
        size: 16
      }), cur),
      direction: 'desc',
      attribute: cur
    }];
    return [].concat((0, _toConsumableArray2.default)(acc), opts);
  }, []) : [];

  // handlers
  var sortHandler = function sortHandler(_ref2) {
    var direction = _ref2.direction,
      attribute = _ref2.attribute;
    setSortAttribute(attribute);
    setSortDirection(direction);
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: blockClass
  }, sortByOpts.length > 0 && /*#__PURE__*/_react.default.createElement(_react2.OverflowMenu, {
    renderIcon: function renderIcon(props) {
      return /*#__PURE__*/_react.default.createElement(_icons.ArrowsVertical, (0, _extends2.default)({
        size: 32
      }, props));
    },
    className: "".concat(blockClass, "_overflow"),
    flipped: true,
    "aria-label": sortByLabel,
    iconDescription: sortByLabel
  }, sortByOpts.map(function (opt) {
    return /*#__PURE__*/_react.default.createElement(_react2.OverflowMenuItem, {
      className: "".concat(blockClass, "_overflow-item"),
      key: opt.id,
      itemText: opt.itemText,
      onClick: function onClick() {
        return sortHandler(opt);
      }
    });
  })));
};
exports.AddSelectSort = AddSelectSort;
AddSelectSort.propTypes = {
  setSortAttribute: _propTypes.default.func,
  setSortDirection: _propTypes.default.func,
  sortBy: _propTypes.default.array,
  sortByLabel: _propTypes.default.string
};
AddSelectSort.displayName = componentName;
