// @carbon/ibm-products/lib/components/AddSelect/AddSelectList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AddSelectList = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _classnames = _interopRequireDefault(require("classnames"));
var _settings = require("../../settings");
var _AddSelectRow = require("./AddSelectRow");
var _useFocus3 = _interopRequireDefault(require("./hooks/useFocus"));
var _excluded = ["filteredItems", "multi"]; //
// Copyright IBM Corp. 2022, 2022
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//
var blockClass = "".concat(_settings.pkg.prefix, "--add-select__selections");
var componentName = 'AddSelectList';
var AddSelectList = function AddSelectList(_ref) {
  var filteredItems = _ref.filteredItems,
    multi = _ref.multi,
    rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var _useFocus = (0, _useFocus3.default)(filteredItems.length),
    _useFocus2 = (0, _slicedToArray2.default)(_useFocus, 2),
    focus = _useFocus2[0],
    setFocus = _useFocus2[1];
  return /*#__PURE__*/_react.default.createElement("div", {
    id: "add-select-focus",
    className: (0, _classnames.default)("".concat(blockClass, "-wrapper"), (0, _defineProperty2.default)({}, "".concat(blockClass, "-wrapper-multi"), multi))
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass)
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "-body"),
    role: "treegrid",
    "aria-label": "add select tree label"
  }, filteredItems.map(function (item, index) {
    return /*#__PURE__*/_react.default.createElement(_AddSelectRow.AddSelectRow, (0, _extends2.default)({
      key: item.id,
      index: index,
      focus: focus,
      setSize: filteredItems.length,
      item: item,
      multi: multi,
      setFocus: setFocus
    }, rest));
  }))));
};
exports.AddSelectList = AddSelectList;
AddSelectList.propTypes = {
  filteredItems: _propTypes.default.array,
  focus: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  multi: _propTypes.default.bool
};
AddSelectList.displayName = componentName;
