// @carbon/ibm-products/lib/components/AddSelect/AddSelectBreadcrumbs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AddSelectBreadcrumbs = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@carbon/react");
var _propTypes = _interopRequireDefault(require("prop-types"));
var _classnames = _interopRequireDefault(require("classnames"));
var _settings = require("../../settings");
//
// Copyright IBM Corp. 2022, 2022
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

var blockClass = "".concat(_settings.pkg.prefix, "--add-select__breadcrumbs");
var componentName = 'AddSelectBreadcrumbs';
var AddSelectBreadcrumbs = function AddSelectBreadcrumbs(_ref) {
  var multi = _ref.multi,
    onClick = _ref.onClick,
    path = _ref.path;
  var clickHandler = function clickHandler(e, idx) {
    e.preventDefault();
    onClick(idx);
  };
  var classNames = (0, _classnames.default)(blockClass, (0, _defineProperty2.default)({}, "".concat(blockClass, "-multi"), multi));
  return /*#__PURE__*/_react.default.createElement(_react2.Breadcrumb, {
    noTrailingSlash: true,
    className: classNames
  }, path.map(function (entry, idx) {
    var isCurrentPage = idx === path.length - 1;
    return /*#__PURE__*/_react.default.createElement(_react2.BreadcrumbItem, {
      key: entry.id,
      isCurrentPage: isCurrentPage,
      onClick: function onClick(e) {
        return clickHandler(e, idx);
      },
      href: "#"
    }, entry.title);
  }));
};
exports.AddSelectBreadcrumbs = AddSelectBreadcrumbs;
AddSelectBreadcrumbs.propTypes = {
  multi: _propTypes.default.bool,
  onClick: _propTypes.default.func,
  path: _propTypes.default.array
};
AddSelectBreadcrumbs.displayName = componentName;
