// @carbon/ibm-products/lib/components/FilterSummary/FilterSummary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("@carbon/react");
var _react2 = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _classnames = _interopRequireDefault(require("classnames"));
var _TagSet = require("../TagSet");
var _settings = require("../../settings");
/**
 * Copyright IBM Corp. 2022, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

var blockClass = "".concat(_settings.pkg.prefix, "--filter-summary");
var FilterSummary = /*#__PURE__*/_react2.default.forwardRef(function (_ref, ref) {
  var _ref$className = _ref.className,
    className = _ref$className === void 0 ? '' : _ref$className,
    _ref$clearFiltersText = _ref.clearFiltersText,
    clearFiltersText = _ref$clearFiltersText === void 0 ? 'Clear filters' : _ref$clearFiltersText,
    _ref$clearFilters = _ref.clearFilters,
    clearFilters = _ref$clearFilters === void 0 ? function () {} : _ref$clearFilters,
    _ref$filters = _ref.filters,
    filters = _ref$filters === void 0 ? [] : _ref$filters;
  var tagFilters = filters.map(function (_ref2) {
    var key = _ref2.key,
      value = _ref2.value;
    return {
      type: 'gray',
      label: "".concat(key, ": ").concat(value)
    };
  });
  return /*#__PURE__*/_react2.default.createElement("div", {
    ref: ref,
    className: (0, _classnames.default)([blockClass, className])
  }, /*#__PURE__*/_react2.default.createElement(_TagSet.TagSet, {
    allTagsModalSearchLabel: "Search all tags",
    allTagsModalSearchPlaceholderText: "Search all tags",
    allTagsModalTitle: "All tags",
    showAllTagsLabel: "View all tags",
    tags: tagFilters
  }), /*#__PURE__*/_react2.default.createElement(_react.Button, {
    kind: "ghost",
    size: "sm",
    onClick: clearFilters
  }, clearFiltersText));
});
var componentName = 'FilterSummary';
FilterSummary.displayName = componentName;
FilterSummary.propTypes = {
  className: _propTypes.default.string,
  clearFilters: _propTypes.default.func.isRequired,
  clearFiltersText: _propTypes.default.string,
  filters: _propTypes.default.arrayOf(_propTypes.default.object).isRequired
};
var _default = FilterSummary;
exports.default = _default;
