// @carbon/ibm-products/lib/components/ActionSet/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.actionsOptions = exports.actionsMapping = exports.actionsLabels = void 0;
/**
 * Copyright IBM Corp. 2021, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

var actionsOptions = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
exports.actionsOptions = actionsOptions;
var actionsLabels = {
  0: 'None',
  1: 'One button',
  2: 'A danger button',
  3: 'A ghost button',
  4: 'Two buttons',
  5: 'Two buttons with one ghost',
  6: 'Three buttons',
  7: 'Three buttons with one ghost',
  8: 'Three buttons with one danger',
  9: 'Four buttons with one ghost',
  10: 'Four buttons with two danger'
};
exports.actionsLabels = actionsLabels;
var actionsMapping = function actionsMapping(labels, action) {
  var _labels$primary, _labels$danger, _labels$secondary, _labels$secondary2, _labels$dangerGhost, _labels$ghost;
  var act = function act(label, kind, key) {
    var actionCall = action && action("Click on '".concat(label, "'"));
    return {
      label: label,
      kind: kind,
      key: key,
      onClick: actionCall && function (evt) {
        evt && evt.persist();
        actionCall(evt);
      }
    };
  };
  var primary = act((_labels$primary = labels === null || labels === void 0 ? void 0 : labels.primary) !== null && _labels$primary !== void 0 ? _labels$primary : 'Primary', 'primary', 1);
  var danger = act((_labels$danger = labels === null || labels === void 0 ? void 0 : labels.danger) !== null && _labels$danger !== void 0 ? _labels$danger : 'Danger', 'danger', 2);
  var secondary = act((_labels$secondary = labels === null || labels === void 0 ? void 0 : labels.secondary) !== null && _labels$secondary !== void 0 ? _labels$secondary : 'Secondary', 'secondary', 3);
  var secondary2 = act((_labels$secondary2 = labels === null || labels === void 0 ? void 0 : labels.secondary2) !== null && _labels$secondary2 !== void 0 ? _labels$secondary2 : 'Secondary', 'secondary', 4);
  var dangerGhost = act((_labels$dangerGhost = labels === null || labels === void 0 ? void 0 : labels.dangerGhost) !== null && _labels$dangerGhost !== void 0 ? _labels$dangerGhost : 'Danger-ghost', 'danger--ghost', 5);
  var ghost = act((_labels$ghost = labels === null || labels === void 0 ? void 0 : labels.ghost) !== null && _labels$ghost !== void 0 ? _labels$ghost : 'Ghost', 'ghost', 6);
  return {
    0: [],
    1: [primary],
    2: [danger],
    3: [ghost],
    4: [primary, secondary],
    5: [primary, ghost],
    6: [primary, secondary, secondary2],
    7: [primary, secondary, ghost],
    8: [danger, secondary, ghost],
    9: [primary, secondary, secondary2, ghost],
    10: [danger, secondary, secondary2, dangerGhost]
  };
};
exports.actionsMapping = actionsMapping;
