// @carbon/ibm-products/es/components/Datagrid/useColumnCenterAlign.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
/*
 * Licensed Materials - Property of IBM
 * 5724-Q36
 * (c) Copyright IBM Corp. 2020
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 */
import React from 'react';
import cx from 'classnames';
import { pkg } from '../../settings';
var blockClass = "".concat(pkg.prefix, "--datagrid");
var useColumnCenterAlign = function useColumnCenterAlign(hooks) {
  var centerAlignRenderer = function centerAlignRenderer(tableProps, column) {
    return /*#__PURE__*/React.createElement("div", {
      className: cx("".concat(blockClass, "__center-align-cell-renderer"), {
        sortDisabled: !tableProps.isTableSortable || tableProps.column.disableSortBy === true
      })
    }, column.Cell(tableProps));
  };
  var centerAlignHeader = function centerAlignHeader(headerProp, column) {
    return /*#__PURE__*/React.createElement("div", {
      className: "".concat(blockClass, "__center-align-header")
    }, typeof column.Header === 'function' ? column.Header(headerProp) : column.Header);
  };
  var centerAlignedColumns = function centerAlignedColumns(columns) {
    var columnsWithDefaultCells = columns.map(function (column) {
      return _objectSpread(_objectSpread({}, column), {}, {
        Cell: column.centerAlignedColumn ? function (tableProp) {
          return centerAlignRenderer(tableProp, column);
        } : column.Cell,
        Header: column.centerAlignedColumn ? function (headerProp) {
          return centerAlignHeader(headerProp, column);
        } : column.Header
      });
    });
    return _toConsumableArray(columnsWithDefaultCells);
  };
  hooks.visibleColumns.push(centerAlignedColumns);
};
export default useColumnCenterAlign;
