// @carbon/ibm-products/es/components/Datagrid/useNestedRowExpander.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
/* eslint-disable react/prop-types */
/*
 * Licensed Materials - Property of IBM
 * 5724-Q36
 * (c) Copyright IBM Corp. 2020
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 */
import React from 'react';
import { ChevronRight } from '@carbon/react/icons';
import cx from 'classnames';
import { pkg, carbon } from '../../settings';
var blockClass = "".concat(pkg.prefix, "--datagrid");
var useNestedRowExpander = function useNestedRowExpander(hooks) {
  var visibleColumns = function visibleColumns(columns) {
    var expanderColumn = {
      id: 'expander',
      Cell: function Cell(_ref) {
        var _cx;
        var row = _ref.row;
        return row.canExpand && /*#__PURE__*/React.createElement("button", _extends({
          type: "button",
          "aria-label": "Expand current row",
          className: cx("".concat(blockClass, "__row-expander"), "".concat(carbon.prefix, "--btn"), "".concat(carbon.prefix, "--btn--ghost"))
        }, row.getToggleRowExpandedProps()), /*#__PURE__*/React.createElement(ChevronRight, {
          className: cx("".concat(blockClass, "__expander-icon"), (_cx = {}, _defineProperty(_cx, "".concat(blockClass, "__expander-icon--not-open"), !row.isExpanded), _defineProperty(_cx, "".concat(blockClass, "__expander-icon--open"), row.isExpanded), _cx))
        }));
      },
      width: 48,
      disableResizing: true,
      disableSortBy: true,
      Header: ''
    };
    return [expanderColumn].concat(_toConsumableArray(columns));
  };
  hooks.visibleColumns.push(visibleColumns);
};
export default useNestedRowExpander;
