// @carbon/ibm-products/es/components/Datagrid/useRowExpander.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _extends from "@babel/runtime/helpers/extends";
/* eslint-disable react/prop-types */
/*
 * Licensed Materials - Property of IBM
 * 5724-Q36
 * (c) Copyright IBM Corp. 2020
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 */
import React from 'react';
import { ChevronDown, ChevronUp } from '@carbon/react/icons';
import { pkg, carbon } from '../../settings';
import cx from 'classnames';
var blockClass = "".concat(pkg.prefix, "--datagrid");
var useRowExpander = function useRowExpander(hooks) {
  var visibleColumns = function visibleColumns(columns) {
    var expanderColumn = {
      id: 'expander',
      Cell: function Cell(_ref) {
        var row = _ref.row;
        return row.canExpand && /*#__PURE__*/React.createElement("button", _extends({
          type: "button",
          "aria-label": "Expand current row",
          className: cx("".concat(blockClass, "__row-expander"), "".concat(carbon.prefix, "--btn"), "".concat(carbon.prefix, "--btn--ghost"))
        }, row.getToggleRowExpandedProps()), row.isExpanded ? /*#__PURE__*/React.createElement(ChevronUp, {
          className: "".concat(blockClass, "__row-expander--icon")
        }) : /*#__PURE__*/React.createElement(ChevronDown, {
          className: "".concat(blockClass, "__row-expander--icon")
        }));
      },
      width: 48,
      disableResizing: true,
      disableSortBy: true,
      Header: ''
    };
    return [expanderColumn].concat(_toConsumableArray(columns));
  };
  hooks.visibleColumns.push(visibleColumns);
};
export default useRowExpander;
