// @carbon/ibm-products/lib/components/_Canary/Canary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Canary = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _classnames = _interopRequireDefault(require("classnames"));
var _packageSettings = _interopRequireDefault(require("../../global/js/package-settings"));
var _react2 = require("@carbon/react");
var _excluded = ["className", "componentName"];
/**
 * Copyright IBM Corp. 2020, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
// Import portions of React that are needed.
// Other standard imports.
// load the package settings direct, because Canary is used by settings.js
// Carbon and package components we use.
// The block part of our conventional BEM class names (blockClass__E--M).
var blockClass = "".concat(_packageSettings.default.prefix, "--canary");

/**
 *  Canary component used when the component requested is not yet production
 */
var Canary = function Canary(_ref /*, originalArgs*/) {
  var className = _ref.className,
    componentName = _ref.componentName,
    rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var instructions = "\nimport { pkg } from '@carbon/ibm-products';\n// NOTE: must happen before component is first used\npkg.component.".concat(componentName, " = true;\n");
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, rest, {
    className: (0, _classnames.default)(blockClass, className)
  }), /*#__PURE__*/_react.default.createElement("h2", null, "This component ", /*#__PURE__*/_react.default.createElement("strong", null, componentName), " is not ready yet."), /*#__PURE__*/_react.default.createElement("p", null, "To enable it, initialize package flags before the component is first used."), /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement("p", null, "e.g. in main.js"), /*#__PURE__*/_react.default.createElement(_react2.CodeSnippet, {
    type: "multi"
  }, instructions), /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement("p", null, "View a live example on", ' ', /*#__PURE__*/_react.default.createElement("a", {
    href: "https://codesandbox.io/s/example-component-olif5?file=/src/config.js"
  }, "codesandbox"), "."));
};
exports.Canary = Canary;
Canary.propTypes = {
  /** Provide an optional class to be applied to the containing node */
  className: _propTypes.default.string,
  /** Name of the component that is not ready yet */
  componentName: _propTypes.default.string.isRequired
};
