// @carbon/ibm-products/lib/components/Datagrid/useActionsColumn.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _classnames = _interopRequireDefault(require("classnames"));
var _react2 = require("@carbon/react");
var _settings = require("../../settings");
var _excluded = ["id", "itemText", "onClick", "icon", "shouldHideMenuItem"],
  _excluded2 = ["id", "onClick", "shouldHideMenuItem", "shouldDisableMenuItem", "disabled"];
/**
 * Copyright IBM Corp. 2021, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var blockClass = "".concat(_settings.pkg.prefix, "--datagrid");
var useActionsColumn = function useActionsColumn(hooks) {
  (0, _react.useEffect)(function () {
    _settings.pkg.checkReportFeatureEnabled('Datagrid.useActionsColumn');
  }, []);
  var useAttachActionsOnInstance = function useAttachActionsOnInstance(instance) {
    var rowActions = instance.rowActions,
      isFetching = instance.isFetching,
      selectedFlatRows = instance.selectedFlatRows;
    if (rowActions && Array.isArray(rowActions)) {
      var addActionsMenu = function addActionsMenu(props, cellData) {
        var cell = cellData.cell;
        var row = cell.row,
          column = cell.column;
        if (column.isAction) {
          var _cx3;
          return [props, {
            children: /*#__PURE__*/_react.default.createElement("div", {
              className: "".concat(blockClass, "__actions-column-contents")
            }, isFetching && /*#__PURE__*/_react.default.createElement(_react2.IconSkeleton, {
              className: "".concat(blockClass, "__actions-column-loading")
            }), !isFetching && rowActions.length <= 2 && /*#__PURE__*/_react.default.createElement("div", {
              className: "".concat(blockClass, "_actions-column"),
              style: {
                display: 'flex'
              }
            }, rowActions.map(function (action, index) {
              var id = action.id,
                itemText = action.itemText,
                _onClick = action.onClick,
                icon = action.icon,
                shouldHideMenuItem = action.shouldHideMenuItem,
                rest = (0, _objectWithoutProperties2.default)(action, _excluded);
              var hidden = typeof shouldHideMenuItem === 'function' && shouldHideMenuItem(row);
              if (hidden) {
                return null;
              }
              var selectedRowId = selectedFlatRows === null || selectedFlatRows === void 0 ? void 0 : selectedFlatRows.filter(function (item) {
                return item.id === row.id ? item.id : null;
              });
              return /*#__PURE__*/_react.default.createElement("div", {
                className: (0, _classnames.default)("".concat(blockClass, "__actions-column-button"), (0, _defineProperty2.default)({}, "".concat(blockClass, "__disabled-row-action-button"), selectedFlatRows && selectedFlatRows.length && selectedRowId && selectedRowId.length)),
                key: "".concat(itemText, "__").concat(index)
              }, /*#__PURE__*/_react.default.createElement(_react2.OverflowMenu, (0, _extends2.default)({}, rest, {
                renderIcon: icon,
                hasIconOnly: true,
                light: true,
                iconDescription: itemText,
                kind: "ghost",
                className: (0, _classnames.default)((0, _defineProperty2.default)({}, "".concat(blockClass, "__disabled-row-action"), selectedFlatRows && selectedFlatRows.length && selectedRowId && selectedRowId.length)),
                onClick: function onClick(e) {
                  if (selectedFlatRows && selectedFlatRows.length && selectedRowId && selectedRowId.length) {
                    // Row actions should be disabled if row is selected and batchActions toolbar is active
                    return;
                  }
                  e.stopPropagation();
                  _onClick(id, row, e);
                }
              })));
            })), !isFetching && rowActions.length > 2 && /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_react2.OverflowMenu, {
              size: "sm",
              light: true,
              flipped: true,
              onClick: function onClick(e) {
                e.stopPropagation();
              }
            }, rowActions.map(function (action) {
              var id = action.id,
                _onClick2 = action.onClick,
                shouldHideMenuItem = action.shouldHideMenuItem,
                shouldDisableMenuItem = action.shouldDisableMenuItem,
                disabled = action.disabled,
                rest = (0, _objectWithoutProperties2.default)(action, _excluded2);
              var hidden = typeof shouldHideMenuItem === 'function' && shouldHideMenuItem(row);
              // shouldDisableMenuItem will override disabled because it's more specific
              // if shouldDisableMenuItem doesn't exists, fall back to disabled
              var isDisabledByRow = typeof shouldDisableMenuItem === 'function' ? shouldDisableMenuItem(row) : disabled;
              if (hidden) {
                return null;
              }
              return /*#__PURE__*/_react.default.createElement(_react2.OverflowMenuItem, (0, _extends2.default)({}, rest, {
                disabled: isDisabledByRow,
                onClick: function onClick(e) {
                  e.stopPropagation();
                  _onClick2(id, row, e);
                },
                key: id
              }));
            })))),
            className: (0, _classnames.default)((_cx3 = {}, (0, _defineProperty2.default)(_cx3, "".concat(blockClass, "__actions-column-cell"), true), (0, _defineProperty2.default)(_cx3, "".concat(blockClass, "__cell"), true), _cx3)),
            style: {
              width: rowActions.length > 2 ? 48 : 96
            }
          }];
        }
        return [props];
      };
      hooks.getCellProps.push(addActionsMenu);
    }
  };
  var useStickyHeaderWidth = function useStickyHeaderWidth(instance) {
    var rowActions = instance.rowActions;
    if (rowActions && Array.isArray(rowActions)) {
      var addHeaderWidth = function addHeaderWidth(props, cellData) {
        var column = cellData.column;
        if (column.isAction) {
          return [props, {
            style: _objectSpread(_objectSpread({}, props.style), {}, {
              width: rowActions.length > 2 ? 48 : 96 // set header width based on action length
            })
          }];
        }

        return [props];
      };
      hooks.getHeaderProps.push(addHeaderWidth);
    }
  };
  hooks.useInstance.push(useStickyHeaderWidth);
  hooks.useInstance.push(useAttachActionsOnInstance);
};
var _default = useActionsColumn;
exports.default = _default;
