// @carbon/ibm-products/lib/components/Datagrid/useInfiniteScroll.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _debounce = _interopRequireDefault(require("lodash/debounce"));
var _useParentDimensions = _interopRequireDefault(require("./useParentDimensions"));
var _useResizeTable = _interopRequireDefault(require("./useResizeTable"));
var _settings = require("../../settings");
/*
 * Licensed Materials - Property of IBM
 * 5724-Q36
 * (c) Copyright IBM Corp. 2020
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 */

var useInfiniteScroll = function useInfiniteScroll(hooks) {
  (0, _useParentDimensions.default)(hooks);
  (0, _useResizeTable.default)(hooks);
  (0, _react.useEffect)(function () {
    _settings.pkg.checkReportFeatureEnabled('Datagrid.useInfiniteScroll');
  }, []);
  var useInstance = function useInstance(instance) {
    var isFetching = instance.isFetching,
      tableHeight = instance.tableHeight,
      innerListRef = instance.innerListRef,
      fetchMoreData = instance.fetchMoreData,
      tableId = instance.tableId;
    var tableElement = document.querySelector("#".concat(tableId));
    var totalTableHeight = tableHeight || (tableElement === null || tableElement === void 0 ? void 0 : tableElement.clientHeight);
    var loadMoreThreshold = 200;
    var emptyFetchData = function emptyFetchData() {};
    // eslint-disable-next-line react-hooks/exhaustive-deps
    var fetchMore = (0, _react.useCallback)((0, _debounce.default)(fetchMoreData || emptyFetchData, 3000, {
      leading: true,
      trailing: false
    }), [fetchMoreData]);
    var onScroll = function onScroll(_ref) {
      var scrollDirection = _ref.scrollDirection,
        scrollOffset = _ref.scrollOffset;
      if (innerListRef && innerListRef.current) {
        if (!isFetching && scrollDirection === 'forward' && scrollOffset + totalTableHeight >= innerListRef.current.clientHeight - loadMoreThreshold) {
          if (fetchMoreData) {
            fetchMore();
          }
        }
      }
    };
    Object.assign(instance, {
      onScroll: onScroll,
      withVirtualScroll: true
    });
  };
  hooks.useInstance.push(useInstance);
};
var _default = useInfiniteScroll;
exports.default = _default;
