// @carbon/ibm-products/lib/components/Datagrid/useRowSize.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = require("react");
var _RowSize = require("./Datagrid/addons/RowSize");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; } /*
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * Licensed Materials - Property of IBM
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * 5724-Q36
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * (c) Copyright IBM Corp. 2020, 2021
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * US Government Users Restricted Rights - Use, duplication or disclosure
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * restricted by GSA ADP Schedule Contract with IBM Corp.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          */
var useRowSize = function useRowSize(hooks) {
  var _useState = (0, _react.useState)(''),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    internalRowSize = _useState2[0],
    setRowSize = _useState2[1];
  hooks.useInstance.push(function (instance) {
    var rowSizeProps = instance.rowSizeProps,
      rowSizes = instance.rowSizes,
      rowSize = instance.rowSize,
      onRowSizeChange = instance.onRowSizeChange;
    var _ref = rowSizeProps || {},
      labels = _ref.labels;
    Object.assign(instance, {
      rowSize: internalRowSize || rowSize,
      rowSizeDropdownProps: _objectSpread(_objectSpread({
        rowSizes: rowSizes
      }, labels), {}, {
        selectedOption: internalRowSize || rowSize,
        onChange: function onChange(value) {
          setRowSize(value);
          if (typeof onRowSizeChange === 'function') {
            onRowSizeChange(value);
          }
        }
      }),
      RowSizeDropdown: _RowSize.RowSizeDropdown
    });
  });
  var getTableProps = function getTableProps(props, _ref2) {
    var instance = _ref2.instance;
    var rowSize = instance.rowSize;
    if (!rowSize) {
      return props;
    }
    return [props, {
      size: rowSize
    }];
  };
  hooks.getTableProps.push(getTableProps);
};
var _default = useRowSize;
exports.default = _default;
