// @carbon/ibm-products/lib/components/Datagrid/useRowRenderer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _DatagridRow = _interopRequireDefault(require("./Datagrid/DatagridRow"));
/*
 * Licensed Materials - Property of IBM
 * 5724-Q36
 * (c) Copyright IBM Corp. 2020
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 */

var useRowRenderer = function useRowRenderer(hooks) {
  var useInstance = function useInstance(instance) {
    var rows = instance.rows;
    var addRowRenderer = function addRowRenderer(row) {
      return Object.assign(row, {
        RowRenderer: _DatagridRow.default,
        subRows: (row.subRows || []).map(addRowRenderer)
      });
    };
    var rowsWithRenderer = rows.map(addRowRenderer);
    Object.assign(instance, {
      rows: rowsWithRenderer
    });
  };
  hooks.useInstance.push(useInstance);
};
var _default = useRowRenderer;
exports.default = _default;
