// @carbon/ibm-products/lib/components/Datagrid/useSkeletonRows.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
/*
 * Licensed Materials - Property of IBM
 * 5724-Q36
 * (c) Copyright IBM Corp. 2020
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 */
var useSkeletons = function useSkeletons(hooks) {
  var useInstance = function useInstance(instance) {
    var isFetching = instance.isFetching,
      rows = instance.rows;
    var skeletonRow = function skeletonRow(id) {
      return {
        isSkeleton: true,
        values: 'skeleton',
        id: id
      };
    };
    var rowsWithSkeletons = [].concat((0, _toConsumableArray2.default)(rows), [skeletonRow('skeleton-row-1'), skeletonRow('skeleton-row-2'), skeletonRow('skeleton-row-3')]);
    Object.assign(instance, {
      rows: isFetching ? rowsWithSkeletons : rows
    });
  };
  hooks.useInstance.push(useInstance);
};
var _default = useSkeletons;
exports.default = _default;
