// @carbon/ibm-products/lib/components/WebTerminal/hooks/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useWebTerminal = exports.WebTerminalProvider = exports.WebTerminalContext = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _settings = require("../../../settings");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
//
// Copyright IBM Corp. 2022, 2022
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

var WebTerminalContext = /*#__PURE__*/(0, _react.createContext)();
exports.WebTerminalContext = WebTerminalContext;
var componentName = 'WebTerminalProvider';
var WebTerminalProvider = function WebTerminalProvider(_ref) {
  var children = _ref.children;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    open = _useState2[0],
    setOpen = _useState2[1];
  var openWebTerminal = (0, _react.useCallback)(function () {
    return setOpen(true);
  }, []);
  var closeWebTerminal = (0, _react.useCallback)(function () {
    return setOpen(false);
  }, []);
  var toggleWebTerminal = (0, _react.useCallback)(function () {
    return setOpen(!open);
  }, [open]);
  return /*#__PURE__*/_react.default.createElement(WebTerminalContext.Provider, {
    value: {
      open: open,
      openWebTerminal: openWebTerminal,
      closeWebTerminal: closeWebTerminal,
      toggleWebTerminal: toggleWebTerminal
    }
  }, children);
};

// Return a placeholder if not released and not enabled by feature flag
exports.WebTerminalProvider = WebTerminalProvider;
exports.WebTerminalProvider = WebTerminalProvider = _settings.pkg.checkComponentEnabled(WebTerminalProvider, componentName);
WebTerminalProvider.propTypes = {
  /**
   * Provide your own terminal component as children to show up in the web terminal
   */
  children: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.node), _propTypes.default.node]).isRequired
};

// Custom hook that exposes the provided value from context
var useWebTerminal = function useWebTerminal() {
  return (0, _react.useContext)(WebTerminalContext);
};
exports.useWebTerminal = useWebTerminal;
