// @carbon/ibm-products/es/components/AddSelect/AddSelectRow.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
//
// Copyright IBM Corp. 2022, 2022
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

import React, { useRef, useEffect, useCallback } from 'react';
import { Button, Dropdown, Layer } from '@carbon/react';
import { ChevronRight, View } from '@carbon/react/icons';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { pkg } from '../../settings';
import { AddSelectFormControl } from './AddSelectFormControl';
var blockClass = "".concat(pkg.prefix, "--add-select__selections");
var componentName = 'AddSelectList';
export var AddSelectRow = function AddSelectRow(_ref) {
  var _modifiers$options, _cx;
  var appliedModifiers = _ref.appliedModifiers,
    displayMetalPanel = _ref.displayMetalPanel,
    index = _ref.index,
    focus = _ref.focus,
    item = _ref.item,
    metaIconDescription = _ref.metaIconDescription,
    modifiers = _ref.modifiers,
    multi = _ref.multi,
    multiSelection = _ref.multiSelection,
    navIconDescription = _ref.navIconDescription,
    parentId = _ref.parentId,
    parentSelected = _ref.parentSelected,
    setAppliedModifiers = _ref.setAppliedModifiers,
    setDisplayMetaPanel = _ref.setDisplayMetaPanel,
    setFocus = _ref.setFocus,
    setMultiSelection = _ref.setMultiSelection,
    setParentSelected = _ref.setParentSelected,
    setSingleSelection = _ref.setSingleSelection,
    setSize = _ref.setSize,
    singleSelection = _ref.singleSelection;
  var ref = useRef(null);
  useEffect(function () {
    if (focus === index) {
      ref.current.focus();
    }
  }, [focus, index]);
  var isSelected = function isSelected() {
    if (multi) {
      return multiSelection.includes(item.id);
    }
    return item.id === singleSelection;
  };
  var getTabIndex = function getTabIndex() {
    // on initial load make the first item tabbable
    if (index === 0 && focus === '') {
      return 0;
    }

    // make it so only the last focused item is tabbable
    if (focus === index && focus !== '') {
      return 0;
    }

    // make unfocused items un-tabbable
    return -1;
  };
  var focusHandler = useCallback(function (reset) {
    setFocus(reset ? '' : index);
  }, [setFocus, index]);
  var handleSingleSelection = function handleSingleSelection() {
    setSingleSelection(item.id);
  };
  var onSelectKeyDown = function onSelectKeyDown(_ref2) {
    var key = _ref2.key;
    if (key === 'Enter' || key === ' ') {
      if (multi) {
        handleMultiSelection();
      } else {
        handleSingleSelection();
      }
    } else if (key === 'ArrowRight' && item.children) {
      onNavigateItem();
    }
  };
  var handleMultiSelection = function handleMultiSelection() {
    var checked = isSelected();
    if (!checked) {
      var newValues = [].concat(_toConsumableArray(multiSelection), [item.id]);
      setMultiSelection(newValues);
    } else {
      var _newValues = multiSelection.filter(function (v) {
        return v !== item.id;
      });
      setMultiSelection(_newValues);
    }
  };
  var onNavigateItem = function onNavigateItem() {
    focusHandler(true);
    setParentSelected(item.id, item.title, parentId);
  };
  var modifierHandler = function modifierHandler(id, selectedItem) {
    var modifiersClone = _toConsumableArray(appliedModifiers);
    var modifierIdx = modifiersClone.findIndex(function (item) {
      return item.id === id;
    });
    modifiersClone[modifierIdx] = _defineProperty({
      id: id
    }, modifiers.id, selectedItem);
    setAppliedModifiers(modifiersClone);
  };
  var metaPanelHandler = function metaPanelHandler() {
    if (item.meta) {
      setDisplayMetaPanel(item);
    }
  };
  var isInMetaPanel = function isInMetaPanel(id) {
    return id === (displayMetalPanel === null || displayMetalPanel === void 0 ? void 0 : displayMetalPanel.id);
  };
  var hasModifiers = (modifiers === null || modifiers === void 0 ? void 0 : (_modifiers$options = modifiers.options) === null || _modifiers$options === void 0 ? void 0 : _modifiers$options.length) > 0;
  var tabIndex = getTabIndex();
  var selected = isSelected();
  var expanded = parentSelected === item.id;
  return /*#__PURE__*/React.createElement("div", {
    className: cx("".concat(blockClass, "-row"), (_cx = {}, _defineProperty(_cx, "".concat(blockClass, "-row--selected"), isSelected()), _defineProperty(_cx, "".concat(blockClass, "-row-meta--selected"), isInMetaPanel(item.id)), _defineProperty(_cx, "".concat(blockClass, "-row--active"), expanded), _cx)),
    onKeyDown: onSelectKeyDown,
    tabIndex: tabIndex,
    ref: ref,
    role: "row",
    "aria-selected": selected,
    "aria-posinset": index,
    "aria-setsize": setSize,
    "aria-expanded": expanded
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "-cell"),
    role: "gridcell"
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "-cell-wrapper")
  }, multi ? /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(AddSelectFormControl, {
    type: "checkbox",
    item: item,
    selected: selected,
    onClick: handleMultiSelection
  }), hasModifiers && /*#__PURE__*/React.createElement(Layer, null, /*#__PURE__*/React.createElement(Dropdown, {
    id: "add-select-modifier-".concat(item.id),
    type: "inline",
    items: modifiers.options,
    label: modifiers.label,
    disabled: !isSelected(),
    className: "".concat(blockClass, "-dropdown"),
    initialSelectedItem: item[modifiers.id],
    onChange: function onChange(_ref3) {
      var selectedItem = _ref3.selectedItem;
      return modifierHandler(item.id, selectedItem);
    }
  }))) : /*#__PURE__*/React.createElement(AddSelectFormControl, {
    type: "radio",
    item: item,
    selected: selected,
    onClick: handleSingleSelection
  }), item.children && /*#__PURE__*/React.createElement(Button, {
    className: "".concat(blockClass, "-view-children"),
    renderIcon: function renderIcon(props) {
      return /*#__PURE__*/React.createElement(ChevronRight, _extends({
        size: 16
      }, props));
    },
    iconDescription: navIconDescription,
    tooltipPosition: "left",
    tooltipAlignment: "center",
    hasIconOnly: true,
    onClick: onNavigateItem,
    kind: "ghost",
    size: "sm",
    tabIndex: -1,
    "aria-hidden": true
  }), item.meta && /*#__PURE__*/React.createElement(Button, {
    className: "".concat(blockClass, "-view-meta"),
    renderIcon: function renderIcon(props) {
      return /*#__PURE__*/React.createElement(View, _extends({
        size: 16
      }, props));
    },
    iconDescription: metaIconDescription,
    tooltipPosition: "left",
    tooltipAlignment: "center",
    hasIconOnly: true,
    kind: "ghost",
    size: "sm",
    onClick: metaPanelHandler
  }))));
};
AddSelectRow.propTypes = {
  appliedModifiers: PropTypes.array,
  displayMetalPanel: PropTypes.object,
  filteredItems: PropTypes.array,
  focus: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  index: PropTypes.number,
  item: PropTypes.object,
  metaIconDescription: PropTypes.string,
  modifiers: PropTypes.object,
  multi: PropTypes.bool,
  multiSelection: PropTypes.array,
  navIconDescription: PropTypes.string,
  parentId: PropTypes.string,
  parentSelected: PropTypes.string,
  setAppliedModifiers: PropTypes.func,
  setDisplayMetaPanel: PropTypes.func,
  setFocus: PropTypes.func,
  setMultiSelection: PropTypes.func,
  setParentSelected: PropTypes.func,
  setSingleSelection: PropTypes.func,
  setSize: PropTypes.number,
  singleSelection: PropTypes.string
};
AddSelectRow.displayName = componentName;
