// @carbon/ibm-products/es/components/AddSelect/AddSelectFormControl.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from 'react';
import { RadioButton, RadioButtonChecked, Checkbox, CheckboxCheckedFilled } from '@carbon/react/icons';
import PropTypes from 'prop-types';
import { pkg } from '../../settings';
import { UserProfileImage } from '../UserProfileImage';
var componentName = 'AddSelectFormControl';
var blockClass = "".concat(pkg.prefix, "--add-select__selections");
export var AddSelectFormControl = function AddSelectFormControl(_ref) {
  var item = _ref.item,
    onClick = _ref.onClick,
    selected = _ref.selected,
    type = _ref.type;
  var controlProps = _defineProperty({
    onClick: onClick,
    className: "".concat(blockClass, "-form-control-wrapper ").concat(blockClass, "-form-control-wrapper--").concat(type),
    size: 20
  }, "aria-labelledby", "control-label-".concat(item.id));
  var getCheckbox = function getCheckbox() {
    if (selected) {
      return /*#__PURE__*/React.createElement(CheckboxCheckedFilled, controlProps);
    }
    return /*#__PURE__*/React.createElement(Checkbox, controlProps);
  };
  var getRadio = function getRadio() {
    if (selected) {
      return /*#__PURE__*/React.createElement(RadioButtonChecked, controlProps);
    }
    return /*#__PURE__*/React.createElement(RadioButton, controlProps);
  };
  var getFormControl = function getFormControl() {
    if (!item) {
      return;
    }
    if (type === 'checkbox') {
      return getCheckbox();
    } else {
      return getRadio();
    }
  };
  var getAvatarProps = function getAvatarProps(_ref2) {
    var src = _ref2.src,
      alt = _ref2.alt,
      icon = _ref2.icon,
      backgroundColor = _ref2.backgroundColor,
      theme = _ref2.theme;
    return {
      className: "".concat(blockClass, "-cell-avatar"),
      size: 'lg',
      theme: theme,
      image: src,
      imageDescription: alt,
      icon: icon,
      backgroundColor: backgroundColor
    };
  };
  var getItemIcon = function getItemIcon(_ref3) {
    var Icon = _ref3.icon;
    return /*#__PURE__*/React.createElement(Icon, null);
  };
  return /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "-form-control")
  }, getFormControl(), /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "-form-control-label-wrapper")
  }, item.avatar && /*#__PURE__*/React.createElement(UserProfileImage, getAvatarProps(item.avatar)), item.icon && /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "-cell-icon")
  }, getItemIcon(item)), /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "-form-control-label-text")
  }, /*#__PURE__*/React.createElement("span", {
    className: "".concat(blockClass, "-cell-title"),
    id: "control-label-".concat(item.id)
  }, item.title), item.subtitle && /*#__PURE__*/React.createElement("span", {
    className: "".concat(blockClass, "-cell-subtitle")
  }, item.subtitle))));
};
AddSelectFormControl.propTypes = {
  item: PropTypes.object,
  onClick: PropTypes.func,
  selected: PropTypes.bool,
  type: PropTypes.oneOf(['checkbox', 'radio'])
};
AddSelectFormControl.displayName = componentName;
