// @carbon/ibm-products/lib/components/Datagrid/Datagrid/addons/CustomizeColumns/Columns.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _reactDnd = require("react-dnd");
var _reactDndHtml5Backend = require("react-dnd-html5-backend");
var _react2 = require("@carbon/react");
var _immutabilityHelper = _interopRequireDefault(require("immutability-helper"));
var _settings = require("../../../../../settings");
var _classnames = _interopRequireDefault(require("classnames"));
var _DraggableItemsList = require("./DraggableItemsList");
/**
 * Copyright IBM Corp. 2022, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

var blockClass = "".concat(_settings.pkg.prefix, "--datagrid");
var getNextIndex = function getNextIndex(array, currentIndex, key) {
  var newIndex = -1;
  if (key === 'ArrowUp') {
    newIndex = currentIndex - 1 >= 0 ? currentIndex - 1 : array.length - 1;
  }
  if (key === 'ArrowDown') {
    newIndex = currentIndex + 1 < array.length ? currentIndex + 1 : 0;
  }
  return newIndex;
};
var Columns = function Columns(_ref) {
  var getVisibleColumnsCount = _ref.getVisibleColumnsCount,
    filterString = _ref.filterString,
    columns = _ref.columns,
    setColumnsObject = _ref.setColumnsObject,
    onSelectColumn = _ref.onSelectColumn,
    assistiveTextInstructionsLabel = _ref.assistiveTextInstructionsLabel,
    assistiveTextDisabledInstructionsLabel = _ref.assistiveTextDisabledInstructionsLabel,
    selectAllLabel = _ref.selectAllLabel,
    isTableSortable = _ref.isTableSortable;
  var _React$useState = _react.default.useState(''),
    _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
    ariaRegionText = _React$useState2[0],
    setAriaRegionText = _React$useState2[1];
  var _React$useState3 = _react.default.useState(-1),
    _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
    focusIndex = _React$useState4[0],
    setFocusIndex = _React$useState4[1];
  var moveElement = _react.default.useCallback(function (dragIndex, hoverIndex) {
    var dragCard = columns[dragIndex];
    setColumnsObject((0, _immutabilityHelper.default)(columns, {
      $splice: [[dragIndex, 1], [hoverIndex, 0, dragCard]]
    }));
  }, [columns, setColumnsObject]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "__customize-columns-column-list")
  }, /*#__PURE__*/_react.default.createElement(_reactDnd.DndProvider, {
    backend: _reactDndHtml5Backend.HTML5Backend
  }, /*#__PURE__*/_react.default.createElement("ol", {
    className: "".concat(blockClass, "__customize-columns-column-list--focus"),
    role: "listbox",
    "aria-describedby": "".concat(blockClass, "__customize-columns--instructions"),
    onKeyDown: function onKeyDown(e) {
      var nextIndex = getNextIndex(columns, focusIndex, e.key);
      if (nextIndex >= 0) {
        setFocusIndex(nextIndex);
        e.preventDefault();
        e.stopPropagation();
      }
    },
    tabIndex: 0
  }, /*#__PURE__*/_react.default.createElement("span", {
    "aria-live": "assertive",
    className: "".concat(blockClass, "__shared-ui--assistive-text")
  }, ariaRegionText), /*#__PURE__*/_react.default.createElement("span", {
    id: "".concat(blockClass, "__customize-columns--instructions"),
    className: "".concat(blockClass, "__shared-ui--assistive-text")
  }, filterString.length === 0 ? assistiveTextInstructionsLabel : assistiveTextDisabledInstructionsLabel), /*#__PURE__*/_react.default.createElement("div", {
    id: "".concat(blockClass, "__customize-columns-select-all"),
    className: (0, _classnames.default)("".concat(blockClass, "__customize-columns-select-all"), (0, _defineProperty2.default)({}, "".concat(blockClass, "__customize-columns-select-all--selected"), getVisibleColumnsCount() > 0)),
    selected: getVisibleColumnsCount() > 0
  }, /*#__PURE__*/_react.default.createElement(_react2.Checkbox, {
    className: "".concat(blockClass, "__customize-columns-checkbox-wrapper"),
    checked: getVisibleColumnsCount() === columns.length,
    indeterminate: getVisibleColumnsCount() < columns.length && getVisibleColumnsCount() > 0,
    onChange: function onChange() {
      onSelectColumn(columns, getVisibleColumnsCount() !== columns.length);
    },
    id: "".concat(blockClass, "__customization-column-select-all"),
    labelText: selectAllLabel
  })), /*#__PURE__*/_react.default.createElement(_DraggableItemsList.DraggableItemsList, {
    columns: columns,
    filterString: filterString,
    focusIndex: focusIndex,
    getNextIndex: getNextIndex,
    isTableSortable: isTableSortable,
    moveElement: moveElement,
    onSelectColumn: onSelectColumn,
    setAriaRegionText: setAriaRegionText,
    setColumnsObject: setColumnsObject,
    setFocusIndex: setFocusIndex
  }))));
};
Columns.propTypes = {
  assistiveTextDisabledInstructionsLabel: _propTypes.default.string,
  assistiveTextInstructionsLabel: _propTypes.default.string,
  columns: _propTypes.default.array.isRequired,
  disabledInstructionsLabel: _propTypes.default.string,
  filterString: _propTypes.default.string.isRequired,
  getVisibleColumnsCount: _propTypes.default.func.isRequired,
  instructionsLabel: _propTypes.default.string,
  isTableSortable: _propTypes.default.bool.isRequired,
  onSelectColumn: _propTypes.default.func.isRequired,
  selectAllLabel: _propTypes.default.string,
  setColumnStatus: _propTypes.default.func,
  setColumnsObject: _propTypes.default.func.isRequired
};
var _default = Columns;
exports.default = _default;
