// @carbon/ibm-products/lib/components/WebTerminal/WebTerminal.docs-page.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _StoryDocsPage = require("../../global/js/utils/StoryDocsPage");
var stories = _interopRequireWildcard(require("./WebTerminal.stories"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var DocsPage = function DocsPage() {
  return /*#__PURE__*/_react.default.createElement(_StoryDocsPage.StoryDocsPage, {
    blocks: [{
      description: "To get started using the WebTerminal, you have to first wrap your\n              application with the `<WebTerminalProvider>`\n              component. This will allow all the WebTerminal components to be\n              used with the `useWebTerminal()` custom hook.\n          ",
      title: 'Getting Started',
      source: {
        code: "import { WebTerminalProvider } from '@carbon/ibm-products';\n\nconst App = () => {\n  return (\n    <WebTerminalProvider>\n      ...\n    <WebTerminalProvider>\n  )\n};"
      }
    }, {
      title: 'useWebTerminal()',
      source: {
        code: "const { open, openWebTerminal, closeWebTerminal, toggleWebTerminal } =\n  useWebTerminal();"
      }
    }, {
      description: "The `useWebTerminal()` hook returns an object with 4 properties:\n1. Read the `open` state (true or false)\n2. `openWebTerminal` sets the `open` state to true\n3. `closeWebTerminal` sets the `open` state to false\n4. `toggleWebTerminal` toggles between true and false\n\nExample:\n",
      source: {
        code: "const MyComponent = () => {\n  const { open, openWebTerminal, closeWebTerminal, toggleWebTerminal } =\n    useWebTerminal();\n\n  return (\n    <div>\n      <button onClick={() => openWebTerminal()}>Open web terminal</button>\n      <button onClick={() => closeWebTerminal()}>Close web terminal</button>\n      <button onClick={() => toggleWebTerminal()}>Toggle web terminal</button>\n      <p>The web terminal is now {open}</p>\n\n      <WebTerminal>...</WebTerminal>\n    </div>\n  );\n};"
      }
    }, {
      subTitle: 'Passing in your own terminal.',
      description: "The `WebTerminal` is wrapper component for your own\nterminal component that you may have been using for your product.\nSimply pass your own terminal component as children.\n            "
    }, {
      subTitle: 'Optional wrapper component.',
      description: "Optionally you can import `WebTerminalContentWrapper` to have\n              content adjust their width based off the `WebTerminal` width when\n              it is opened.",
      source: {
        code: "<WebTerminalContentWrapper>\n  content\n</WebTerminalContentWrapper>\n\n<WebTerminal>\n  ...\n</WebTerminal>"
      }
    }, {
      description: 'The default web terminal is one which opens and the terminal loads. Then if the user desires to dismiss it they can click the close icon on the top right corner of the component.',
      story: stories.Default
    }, {
      story: stories.WithDocumentationLinks
    }, {
      subTitle: 'Help users with helpful documentation links',
      description: "Optionally you can pass documentation links to help point your users to useful\ndocumentation for your terminal, or kubernetes commands etc.\n\nThe documentation links should reflect the same props as `OverflowMenuItem` from\ncarbon.\n\nLearn more:\nhttps://react.carbondesignsystem.com/?path=/docs/components-overflowmenu--default\n",
      source: {
        code: "const documentationLinks = [\n  {\n    itemText: 'BX/ICP docs',\n    href: '#',\n    onClick: () => console.log('clicked'),\n  },\n  {\n    itemText: 'Kubernetes docs',\n    href: '#',\n    onClick: () => console.log('clicked'),\n  },\n  {\n    itemText: 'Docker docs',\n    href: '#',\n    onClick: () => console.log('clicked'),\n  },\n  {\n    itemText: 'Helm docs',\n    href: '#',\n    onClick: () => console.log('clicked'),\n  },\n];"
      }
    }, {
      story: stories.WithActions
    }, {
      description: "Optionally you can add actions to the web terminal header bar. To add actions\nyou need to pass in an array of actions with objects for each object.",
      source: {
        code: "<WebTerminal\n  actions={[\n    {\n      renderIcon: Code32,\n      onClick: () => {},\n      iconDescription: 'Create new deployment',\n    },\n  ]}\n>\n  ...\n</WebTerminal>"
      }
    }]
  });
};
var _default = DocsPage;
exports.default = _default;
