// @carbon/ibm-products/lib/components/MultiAddSelect/MultiAddSelect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MultiAddSelect = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _AddSelect = require("../AddSelect");
var _devtools = require("../../global/js/utils/devtools");
var _settings = require("../../settings");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
//
// Copyright IBM Corp. 2022, 2022
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

var componentName = 'MultiAddSelect';

/**
 * Used to add or select multiple or more items from larger lists or hierarchies.
 */
var MultiAddSelect = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  return /*#__PURE__*/_react.default.createElement(_AddSelect.AddSelect, (0, _extends2.default)({}, props, {
    multi: true,
    ref: ref
  }, (0, _devtools.getDevtoolsProps)(componentName)));
});
exports.MultiAddSelect = MultiAddSelect;
exports.MultiAddSelect = MultiAddSelect = _settings.pkg.checkComponentEnabled(MultiAddSelect, componentName);
MultiAddSelect.propTypes = {
  /**
   * optional class name
   */
  className: _propTypes.default.string,
  /**
   * placeholder for column input filter
   */
  columnInputPlaceholder: _propTypes.default.string,
  /**
   * text description that appears under the title
   */
  description: _propTypes.default.string,
  /**
   * options to display in the global filter box. values are generated
   * from the id which should correlate with a specific property in an
   * item entry
   */
  globalFilters: _propTypes.default.arrayOf(_propTypes.default.shape({
    id: _propTypes.default.string,
    label: _propTypes.default.string
  })),
  globalFiltersIconDescription: _propTypes.default.string,
  /**
   * placeholder text for the global filter dropdown
   */
  globalFiltersPlaceholderText: _propTypes.default.string,
  /**
   * text for the global filter primary button
   */
  globalFiltersPrimaryButtonText: _propTypes.default.string,
  /**
   * text for the global filter secondary button
   */
  globalFiltersSecondaryButtonText: _propTypes.default.string,
  /**
   * label for global search input
   */
  globalSearchLabel: _propTypes.default.string,
  /**
   * placeholder for global search input
   */
  globalSearchPlaceholder: _propTypes.default.string,
  /**
   * the theme for the empty state illustration
   */
  illustrationTheme: _propTypes.default.oneOf(['light', 'dark']),
  /**
   * title that displays in the sidebar / influencer
   */
  influencerTitle: _propTypes.default.string,
  /**
   * object that contains the item data. for more information reference the
   * "Structuring items" section in the docs tab
   */
  items: _propTypes.default.shape({
    modifiers: _propTypes.default.shape({
      label: _propTypes.default.string,
      options: _propTypes.default.array
    }),
    sortBy: _propTypes.default.array,
    filterBy: _propTypes.default.array,
    entries: _propTypes.default.arrayOf(_propTypes.default.shape({
      avatar: _propTypes.default.shape({
        alt: _propTypes.default.string,
        icon: _propTypes.default.func,
        src: _propTypes.default.string,
        theme: _propTypes.default.oneOf(['light', 'dark'])
      }),
      children: _propTypes.default.object,
      icon: _propTypes.default.func,
      id: _propTypes.default.string.isRequired,
      subtitle: _propTypes.default.string,
      title: _propTypes.default.string.isRequired,
      value: _propTypes.default.string.isRequired
    }))
  }),
  /**
   * label that display above the list of items
   */
  itemsLabel: _propTypes.default.string,
  /**
   * text to display when no results are found from the global search
   */
  noResultsDescription: _propTypes.default.string,
  /**
   * title to display when no results are found from the global search
   */
  noResultsTitle: _propTypes.default.string,
  /**
   * text body that displays in the sidebar when nothing is selected
   */
  noSelectionDescription: _propTypes.default.string,
  /**
   * title that displays in the sidebar when nothing is selected
   */
  noSelectionTitle: _propTypes.default.string,
  /**
   * function to call when the close button clicked
   */
  onClose: _propTypes.default.func,
  /**
   * text for close button
   */
  onCloseButtonText: _propTypes.default.string,
  /**
   * function to call when the submit button is clicked. returns a selection
   */
  onSubmit: _propTypes.default.func,
  /**
   * text for the submit button
   */
  onSubmitButtonText: _propTypes.default.string,
  /**
   * specifies if the component is open or not
   */
  open: _propTypes.default.bool,
  /**
   * text that displays when displaying filtered items
   */
  searchResultsTitle: _propTypes.default.string,
  /**
   * header text
   */
  title: _propTypes.default.string
};
MultiAddSelect.displayName = componentName;
