// @carbon/ibm-products/es/components/CreateSidePanel/CreateSidePanel.docs-page.js
import React from 'react';
import { StoryDocsPage } from '../../global/js/utils/StoryDocsPage';
import * as stories from './CreateSidePanel.stories';
var DocsPage = function DocsPage() {
  return /*#__PURE__*/React.createElement(StoryDocsPage, {
    altGuidelinesHref: [{
      href: 'https://pages.github.ibm.com/cdai-design/pal/patterns/creation-flows/usage#side-panel',
      label: 'Carbon create flows side panel usage guidelines'
    }],
    blocks: [{
      story: stories.Default
    }, {
      title: 'Form validation',
      description: "All forms, including that within the `CreateSidePanel` should follow C&CS\nguidelines for form validation.\n\nThis includes the following:\n\n- The `Submit` button in the side panel should be disabled, until all required\n  inputs are filled in and valid\n- All required inputs should _only_ throw an invalid error _after_ the element\n  loses focus\n- All optional form fields should have an `(optional)` text at the end of the\n  input `labelText`. Optional should always be in parentheses\n\nYou can find more information on how to validate your form fields in\n[Carbon's Form usage page](https://www.carbondesignsystem.com/components/form/usage)."
    }]
  });
};
export default DocsPage;
